// @luzmo/embed/luzmo-embed.libs.js
(function (g, f) {
  var hasExports = typeof exports === 'object';
  if (typeof define === "function" && define.amd) {
    define([], f);
  } else if (typeof module === "object" && module.exports) {
    module.exports = f();
  } else {
    var m = hasExports ? f() : f();
    var root = hasExports ? exports : g;
    for(var i in m) root[i] = m[i];
  }}(typeof self !== 'undefined' ? self : this, () => {
var exports = {};
var module = { exports };
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i6 = decorators.length - 1, decorator; i6 >= 0; i6--)
    if (decorator = decorators[i6])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result)
    __defProp(target, key, result);
  return result;
};

// libs/embed-libs/embed/src/index.libs.ts
var index_libs_exports = {};
__export(index_libs_exports, {
  LuzmoDashboardWebComponent: () => LuzmoDashboardWebComponent
});
module.exports = __toCommonJS(index_libs_exports);

// node_modules/@lit/reactive-element/decorators/custom-element.js
var t = (t5) => (e7, o6) => {
  void 0 !== o6 ? o6.addInitializer(() => {
    customElements.define(t5, e7);
  }) : customElements.define(t5, e7);
};

// node_modules/@lit/reactive-element/css-tag.js
var t2 = globalThis;
var e = t2.ShadowRoot && (void 0 === t2.ShadyCSS || t2.ShadyCSS.nativeShadow) && "adoptedStyleSheets" in Document.prototype && "replace" in CSSStyleSheet.prototype;
var s = Symbol();
var o = /* @__PURE__ */ new WeakMap();
var n = class {
  constructor(t5, e7, o6) {
    if (this._$cssResult$ = true, o6 !== s)
      throw Error("CSSResult is not constructable. Use `unsafeCSS` or `css` instead.");
    this.cssText = t5, this.t = e7;
  }
  get styleSheet() {
    let t5 = this.o;
    const s4 = this.t;
    if (e && void 0 === t5) {
      const e7 = void 0 !== s4 && 1 === s4.length;
      e7 && (t5 = o.get(s4)), void 0 === t5 && ((this.o = t5 = new CSSStyleSheet()).replaceSync(this.cssText), e7 && o.set(s4, t5));
    }
    return t5;
  }
  toString() {
    return this.cssText;
  }
};
var r = (t5) => new n("string" == typeof t5 ? t5 : t5 + "", void 0, s);
var i = (t5, ...e7) => {
  const o6 = 1 === t5.length ? t5[0] : e7.reduce((e8, s4, o7) => e8 + ((t6) => {
    if (true === t6._$cssResult$)
      return t6.cssText;
    if ("number" == typeof t6)
      return t6;
    throw Error("Value passed to 'css' function must be a 'css' function result: " + t6 + ". Use 'unsafeCSS' to pass non-literal values, but take care to ensure page security.");
  })(s4) + t5[o7 + 1], t5[0]);
  return new n(o6, t5, s);
};
var S = (s4, o6) => {
  if (e)
    s4.adoptedStyleSheets = o6.map((t5) => t5 instanceof CSSStyleSheet ? t5 : t5.styleSheet);
  else
    for (const e7 of o6) {
      const o7 = document.createElement("style"), n6 = t2.litNonce;
      void 0 !== n6 && o7.setAttribute("nonce", n6), o7.textContent = e7.cssText, s4.appendChild(o7);
    }
};
var c = e ? (t5) => t5 : (t5) => t5 instanceof CSSStyleSheet ? ((t6) => {
  let e7 = "";
  for (const s4 of t6.cssRules)
    e7 += s4.cssText;
  return r(e7);
})(t5) : t5;

// node_modules/@lit/reactive-element/reactive-element.js
var { is: i2, defineProperty: e2, getOwnPropertyDescriptor: r2, getOwnPropertyNames: h, getOwnPropertySymbols: o2, getPrototypeOf: n2 } = Object;
var a = globalThis;
var c2 = a.trustedTypes;
var l = c2 ? c2.emptyScript : "";
var p = a.reactiveElementPolyfillSupport;
var d = (t5, s4) => t5;
var u = { toAttribute(t5, s4) {
  switch (s4) {
    case Boolean:
      t5 = t5 ? l : null;
      break;
    case Object:
    case Array:
      t5 = null == t5 ? t5 : JSON.stringify(t5);
  }
  return t5;
}, fromAttribute(t5, s4) {
  let i6 = t5;
  switch (s4) {
    case Boolean:
      i6 = null !== t5;
      break;
    case Number:
      i6 = null === t5 ? null : Number(t5);
      break;
    case Object:
    case Array:
      try {
        i6 = JSON.parse(t5);
      } catch (t6) {
        i6 = null;
      }
  }
  return i6;
} };
var f = (t5, s4) => !i2(t5, s4);
var y = { attribute: true, type: String, converter: u, reflect: false, hasChanged: f };
Symbol.metadata ??= Symbol("metadata"), a.litPropertyMetadata ??= /* @__PURE__ */ new WeakMap();
var b = class extends HTMLElement {
  static addInitializer(t5) {
    this._$Ei(), (this.l ??= []).push(t5);
  }
  static get observedAttributes() {
    return this.finalize(), this._$Eh && [...this._$Eh.keys()];
  }
  static createProperty(t5, s4 = y) {
    if (s4.state && (s4.attribute = false), this._$Ei(), this.elementProperties.set(t5, s4), !s4.noAccessor) {
      const i6 = Symbol(), r7 = this.getPropertyDescriptor(t5, i6, s4);
      void 0 !== r7 && e2(this.prototype, t5, r7);
    }
  }
  static getPropertyDescriptor(t5, s4, i6) {
    const { get: e7, set: h3 } = r2(this.prototype, t5) ?? { get() {
      return this[s4];
    }, set(t6) {
      this[s4] = t6;
    } };
    return { get() {
      return e7?.call(this);
    }, set(s5) {
      const r7 = e7?.call(this);
      h3.call(this, s5), this.requestUpdate(t5, r7, i6);
    }, configurable: true, enumerable: true };
  }
  static getPropertyOptions(t5) {
    return this.elementProperties.get(t5) ?? y;
  }
  static _$Ei() {
    if (this.hasOwnProperty(d("elementProperties")))
      return;
    const t5 = n2(this);
    t5.finalize(), void 0 !== t5.l && (this.l = [...t5.l]), this.elementProperties = new Map(t5.elementProperties);
  }
  static finalize() {
    if (this.hasOwnProperty(d("finalized")))
      return;
    if (this.finalized = true, this._$Ei(), this.hasOwnProperty(d("properties"))) {
      const t6 = this.properties, s4 = [...h(t6), ...o2(t6)];
      for (const i6 of s4)
        this.createProperty(i6, t6[i6]);
    }
    const t5 = this[Symbol.metadata];
    if (null !== t5) {
      const s4 = litPropertyMetadata.get(t5);
      if (void 0 !== s4)
        for (const [t6, i6] of s4)
          this.elementProperties.set(t6, i6);
    }
    this._$Eh = /* @__PURE__ */ new Map();
    for (const [t6, s4] of this.elementProperties) {
      const i6 = this._$Eu(t6, s4);
      void 0 !== i6 && this._$Eh.set(i6, t6);
    }
    this.elementStyles = this.finalizeStyles(this.styles);
  }
  static finalizeStyles(s4) {
    const i6 = [];
    if (Array.isArray(s4)) {
      const e7 = new Set(s4.flat(1 / 0).reverse());
      for (const s5 of e7)
        i6.unshift(c(s5));
    } else
      void 0 !== s4 && i6.push(c(s4));
    return i6;
  }
  static _$Eu(t5, s4) {
    const i6 = s4.attribute;
    return false === i6 ? void 0 : "string" == typeof i6 ? i6 : "string" == typeof t5 ? t5.toLowerCase() : void 0;
  }
  constructor() {
    super(), this._$Ep = void 0, this.isUpdatePending = false, this.hasUpdated = false, this._$Em = null, this._$Ev();
  }
  _$Ev() {
    this._$ES = new Promise((t5) => this.enableUpdating = t5), this._$AL = /* @__PURE__ */ new Map(), this._$E_(), this.requestUpdate(), this.constructor.l?.forEach((t5) => t5(this));
  }
  addController(t5) {
    (this._$EO ??= /* @__PURE__ */ new Set()).add(t5), void 0 !== this.renderRoot && this.isConnected && t5.hostConnected?.();
  }
  removeController(t5) {
    this._$EO?.delete(t5);
  }
  _$E_() {
    const t5 = /* @__PURE__ */ new Map(), s4 = this.constructor.elementProperties;
    for (const i6 of s4.keys())
      this.hasOwnProperty(i6) && (t5.set(i6, this[i6]), delete this[i6]);
    t5.size > 0 && (this._$Ep = t5);
  }
  createRenderRoot() {
    const t5 = this.shadowRoot ?? this.attachShadow(this.constructor.shadowRootOptions);
    return S(t5, this.constructor.elementStyles), t5;
  }
  connectedCallback() {
    this.renderRoot ??= this.createRenderRoot(), this.enableUpdating(true), this._$EO?.forEach((t5) => t5.hostConnected?.());
  }
  enableUpdating(t5) {
  }
  disconnectedCallback() {
    this._$EO?.forEach((t5) => t5.hostDisconnected?.());
  }
  attributeChangedCallback(t5, s4, i6) {
    this._$AK(t5, i6);
  }
  _$EC(t5, s4) {
    const i6 = this.constructor.elementProperties.get(t5), e7 = this.constructor._$Eu(t5, i6);
    if (void 0 !== e7 && true === i6.reflect) {
      const r7 = (void 0 !== i6.converter?.toAttribute ? i6.converter : u).toAttribute(s4, i6.type);
      this._$Em = t5, null == r7 ? this.removeAttribute(e7) : this.setAttribute(e7, r7), this._$Em = null;
    }
  }
  _$AK(t5, s4) {
    const i6 = this.constructor, e7 = i6._$Eh.get(t5);
    if (void 0 !== e7 && this._$Em !== e7) {
      const t6 = i6.getPropertyOptions(e7), r7 = "function" == typeof t6.converter ? { fromAttribute: t6.converter } : void 0 !== t6.converter?.fromAttribute ? t6.converter : u;
      this._$Em = e7, this[e7] = r7.fromAttribute(s4, t6.type), this._$Em = null;
    }
  }
  requestUpdate(t5, s4, i6) {
    if (void 0 !== t5) {
      if (i6 ??= this.constructor.getPropertyOptions(t5), !(i6.hasChanged ?? f)(this[t5], s4))
        return;
      this.P(t5, s4, i6);
    }
    false === this.isUpdatePending && (this._$ES = this._$ET());
  }
  P(t5, s4, i6) {
    this._$AL.has(t5) || this._$AL.set(t5, s4), true === i6.reflect && this._$Em !== t5 && (this._$Ej ??= /* @__PURE__ */ new Set()).add(t5);
  }
  async _$ET() {
    this.isUpdatePending = true;
    try {
      await this._$ES;
    } catch (t6) {
      Promise.reject(t6);
    }
    const t5 = this.scheduleUpdate();
    return null != t5 && await t5, !this.isUpdatePending;
  }
  scheduleUpdate() {
    return this.performUpdate();
  }
  performUpdate() {
    if (!this.isUpdatePending)
      return;
    if (!this.hasUpdated) {
      if (this.renderRoot ??= this.createRenderRoot(), this._$Ep) {
        for (const [t7, s5] of this._$Ep)
          this[t7] = s5;
        this._$Ep = void 0;
      }
      const t6 = this.constructor.elementProperties;
      if (t6.size > 0)
        for (const [s5, i6] of t6)
          true !== i6.wrapped || this._$AL.has(s5) || void 0 === this[s5] || this.P(s5, this[s5], i6);
    }
    let t5 = false;
    const s4 = this._$AL;
    try {
      t5 = this.shouldUpdate(s4), t5 ? (this.willUpdate(s4), this._$EO?.forEach((t6) => t6.hostUpdate?.()), this.update(s4)) : this._$EU();
    } catch (s5) {
      throw t5 = false, this._$EU(), s5;
    }
    t5 && this._$AE(s4);
  }
  willUpdate(t5) {
  }
  _$AE(t5) {
    this._$EO?.forEach((t6) => t6.hostUpdated?.()), this.hasUpdated || (this.hasUpdated = true, this.firstUpdated(t5)), this.updated(t5);
  }
  _$EU() {
    this._$AL = /* @__PURE__ */ new Map(), this.isUpdatePending = false;
  }
  get updateComplete() {
    return this.getUpdateComplete();
  }
  getUpdateComplete() {
    return this._$ES;
  }
  shouldUpdate(t5) {
    return true;
  }
  update(t5) {
    this._$Ej &&= this._$Ej.forEach((t6) => this._$EC(t6, this[t6])), this._$EU();
  }
  updated(t5) {
  }
  firstUpdated(t5) {
  }
};
b.elementStyles = [], b.shadowRootOptions = { mode: "open" }, b[d("elementProperties")] = /* @__PURE__ */ new Map(), b[d("finalized")] = /* @__PURE__ */ new Map(), p?.({ ReactiveElement: b }), (a.reactiveElementVersions ??= []).push("2.0.4");

// node_modules/@lit/reactive-element/decorators/property.js
var o3 = { attribute: true, type: String, converter: u, reflect: false, hasChanged: f };
var r3 = (t5 = o3, e7, r7) => {
  const { kind: n6, metadata: i6 } = r7;
  let s4 = globalThis.litPropertyMetadata.get(i6);
  if (void 0 === s4 && globalThis.litPropertyMetadata.set(i6, s4 = /* @__PURE__ */ new Map()), s4.set(r7.name, t5), "accessor" === n6) {
    const { name: o6 } = r7;
    return { set(r8) {
      const n7 = e7.get.call(this);
      e7.set.call(this, r8), this.requestUpdate(o6, n7, t5);
    }, init(e8) {
      return void 0 !== e8 && this.P(o6, void 0, t5), e8;
    } };
  }
  if ("setter" === n6) {
    const { name: o6 } = r7;
    return function(r8) {
      const n7 = this[o6];
      e7.call(this, r8), this.requestUpdate(o6, n7, t5);
    };
  }
  throw Error("Unsupported decorator location: " + n6);
};
function n3(t5) {
  return (e7, o6) => "object" == typeof o6 ? r3(t5, e7, o6) : ((t6, e8, o7) => {
    const r7 = e8.hasOwnProperty(o7);
    return e8.constructor.createProperty(o7, r7 ? { ...t6, wrapped: true } : t6), r7 ? Object.getOwnPropertyDescriptor(e8, o7) : void 0;
  })(t5, e7, o6);
}

// node_modules/@lit/reactive-element/decorators/state.js
function r4(r7) {
  return n3({ ...r7, state: true, attribute: false });
}

// node_modules/@lit/reactive-element/decorators/base.js
var e3 = (e7, t5, c4) => (c4.configurable = true, c4.enumerable = true, Reflect.decorate && "object" != typeof t5 && Object.defineProperty(e7, t5, c4), c4);

// node_modules/@lit/reactive-element/decorators/query.js
function e4(e7, r7) {
  return (n6, s4, i6) => {
    const o6 = (t5) => t5.renderRoot?.querySelector(e7) ?? null;
    if (r7) {
      const { get: e8, set: r8 } = "object" == typeof s4 ? n6 : i6 ?? (() => {
        const t5 = Symbol();
        return { get() {
          return this[t5];
        }, set(e9) {
          this[t5] = e9;
        } };
      })();
      return e3(n6, s4, { get() {
        let t5 = e8.call(this);
        return void 0 === t5 && (t5 = o6(this), (null !== t5 || this.hasUpdated) && r8.call(this, t5)), t5;
      } });
    }
    return e3(n6, s4, { get() {
      return o6(this);
    } });
  };
}

// libs/embed-libs/shared-embed/src/lib/helpers/error.ts
var _handleError = (errorMessage, parameters) => {
  let message = "The dashboard component encountered an error";
  if (typeof parameters?.dashboardId === "string") {
    message += ` for dashboard with id ${parameters.dashboardId}`;
  }
  if (typeof parameters?.dashboardSlug === "string") {
    message += ` for dashboard with slug ${parameters.dashboardSlug}`;
  }
  if (typeof parameters?.container === "string") {
    message += ` in container ${parameters.container}`;
  }
  if (errorMessage) {
    message += `:
${errorMessage}`;
    console.warn(message);
  }
};

// libs/embed-libs/shared-embed/src/lib/helpers/browser-check.ts
var _browserCheck = (dashboardComponent) => {
  let hasError = false;
  let errorMessage = "";
  const document_ = document.body || document.documentElement;
  const style = document_.style;
  const browserCheck = {
    svg: !!("createElementNS" in document && document.createElementNS("http://www.w3.org/2000/svg", "svg").createSVGRect),
    flexbox: style.webkitFlexWrap === "" || style.flexWrap === "",
    postMessage: !!window.postMessage
  };
  const browserCheckKeys = ["svg", "flexbox", "postMessage"];
  for (const key of browserCheckKeys) {
    if (!browserCheck[key]) {
      hasError = true;
      dashboardComponent.browserNotSupported = dashboardComponent.browserNotSupported || [];
      dashboardComponent.browserNotSupported.push(key);
      errorMessage += `Your browser does not support: ${key}
`;
    }
  }
  const ua = window.navigator.userAgent;
  const msie = ua.includes("MSIE");
  const trident = ua.includes("Trident/");
  if (msie || trident) {
    hasError = true;
    errorMessage = "IE is not supported.";
  }
  if (hasError) {
    _handleError(errorMessage);
  }
};

// libs/embed-libs/shared-embed/src/lib/helpers/default-parameters.ts
var DEFAULT_PARAMETERS = {
  appServer: "https://app.luzmo.com/",
  apiHost: "https://api.luzmo.com/",
  dashboardId: null,
  dashboardSlug: null,
  container: null,
  key: null,
  token: null,
  language: "auto",
  qeVersion: null,
  editMode: null,
  screenMode: null,
  switchScreenModeOnResize: true,
  itemId: null,
  mainColor: "",
  accentColor: "",
  itemDimensions: {
    width: "auto",
    height: "auto"
  },
  loaderBackground: "#f9f9f9",
  loaderFontColor: "#5a5a5a",
  loaderSpinnerColor: "rgba(255, 165, 0, 0.7)",
  loaderSpinnerBackground: "rgba(169, 169, 169, 0.14)",
  timezoneId: null,
  contextId: null
};

// libs/embed-libs/shared-embed/src/lib/helpers/extents.ts
var EXTENTS = [
  { mode: "mobile", width: [150, 767] },
  { mode: "tablet", width: [768, 1199] },
  { mode: "desktop", width: [1200, 1599] },
  { mode: "largeScreen", width: [1600, 9999] },
  { mode: "fixed", width: [] }
];

// libs/embed-libs/shared-embed/src/lib/helpers/translations.ts
var TRANSLATIONS = {
  initializing: { en: "Initializing...", nl: "Laden...", fr: "Chargement..." },
  errorDashboard: {
    en: "Oops, we could not load this dashboard!",
    nl: "Oeps, we konden het dashboard niet laden.",
    fr: "Le chargement a echou\xE9."
  },
  errorItem: {
    en: "Oops, we could not load this item!",
    nl: "Oeps, we konden de grafiek niet laden.",
    fr: "Le chargement a echou\xE9."
  }
};
var translations_default = TRANSLATIONS;

// libs/embed-libs/shared-embed/src/lib/helpers/utils.ts
var addTrailingSlash = (serverUrl) => {
  if (!serverUrl) {
    return "";
  }
  return serverUrl.endsWith("/") ? serverUrl : `${serverUrl}/`;
};
var removeTrailingSlash = (serverUrl) => {
  if (!serverUrl) {
    return "";
  }
  return serverUrl.endsWith("/") ? serverUrl.slice(0, -1) : serverUrl;
};
var formatDashboardList = (dashboards, integrationId, user, dashboardComponent) => {
  const combineRights = (securable, user2) => {
    const matchingUser = securable.users.find((u3) => u3.id === user2.id);
    const securableAccessRights = matchingUser && matchingUser.userAccessRight ? {
      flagRead: matchingUser.userAccessRight.flagRead,
      flagUse: matchingUser.userAccessRight.flagUse,
      flagModify: matchingUser.userAccessRight.flagModify,
      flagOwn: matchingUser.userAccessRight.flagOwn
    } : {
      flagRead: false,
      flagUse: false,
      flagModify: false,
      flagOwn: false
    };
    const groups = securable.groups.filter(
      (group) => user2.userGroups.includes(group.id) || group.public === true
    );
    const groupsAccess = groups.map((group) => {
      if (group.groupAccessRight) {
        return group.groupAccessRight;
      }
    });
    for (const groupAccess of groupsAccess) {
      if (!securableAccessRights.flagRead && groupAccess.flagRead) {
        securableAccessRights.flagRead = groupAccess.flagRead;
      }
      if (!securableAccessRights.flagUse && groupAccess.flagUse) {
        securableAccessRights.flagUse = groupAccess.flagUse;
      }
      if (!securableAccessRights.flagModify && groupAccess.flagModify) {
        securableAccessRights.flagModify = groupAccess.flagModify;
      }
      if (!securableAccessRights.flagOwn && groupAccess.flagOwn) {
        securableAccessRights.flagOwn = groupAccess.flagOwn;
      }
    }
    const integrationsAccess = securable.integrations?.length > 0 ? securable.integrations.map((integration) => {
      if (integration.integrationAccessRight) {
        return integration.integrationAccessRight;
      }
    }) : [];
    for (const integrationAccess of integrationsAccess) {
      if (!securableAccessRights.flagRead && integrationAccess.flagRead) {
        securableAccessRights.flagRead = integrationAccess.flagRead;
      }
      if (!securableAccessRights.flagUse && integrationAccess.flagUse) {
        securableAccessRights.flagUse = integrationAccess.flagUse;
      }
      if (!securableAccessRights.flagModify && integrationAccess.flagModify) {
        securableAccessRights.flagModify = integrationAccess.flagModify;
      }
    }
    return securableAccessRights;
  };
  const localize = (item, locale) => {
    if (item === null || item === void 0) {
      return "";
    }
    if (typeof item !== "object" || Object.keys(item).length === 0) {
      if (typeof item === "object" && Object.keys(item).length === 0) {
        return "";
      }
      return item;
    }
    let localizedItem;
    localizedItem = locale && item[locale] ? item[locale] : item[Object.keys(item)[0]];
    if (localizedItem === void 0 || localizedItem === null) {
      localizedItem = "";
    }
    return localizedItem;
  };
  const treatedDashboards = [];
  for (const dashboard of dashboards) {
    const treatedDashboard = {
      accessibleBy: [],
      accessRights: combineRights(dashboard, user),
      id: dashboard.id,
      modifiedAt: dashboard.modified_at,
      name: localize(dashboard.name, dashboardComponent.language),
      slug: null,
      tags: dashboard.tags.map((t5) => t5.tag).sort((t1, t22) => t1.localeCompare(t22))
    };
    const currentIntegration = dashboard.integrations.find(
      (i6) => i6.id === integrationId
    );
    if (currentIntegration?.integrationAccessRight?.slug) {
      treatedDashboard.slug = currentIntegration.integrationAccessRight.slug;
    }
    const users = dashboard.users.map((u3) => ({
      model: "User",
      id: u3.id,
      name: localize(u3.name, user.locale_id)
    }));
    const groups = dashboard.groups.map((g2) => ({
      model: "Group",
      id: g2.id,
      name: localize(g2.name, user.locale_id) || {}
    }));
    const integrations = dashboard.integrations.map((i6) => ({
      model: "Integration",
      id: i6.id,
      name: localize(i6.name, user.locale_id) || {}
    }));
    treatedDashboard.accessibleBy = [...users, ...groups, ...integrations];
    treatedDashboards.push(treatedDashboard);
  }
  return treatedDashboards;
};
var VERSION = "0.0.0-shared-luzmo";
var sanitizeUrl = (url) => {
  const invalidPrototcolRegex = /^(%20|\s)*(javascript|data)/im;
  const ctrlCharactersRegex = /[^\u0020-\u007E]/gim;
  const urlSchemeRegex = /^([^:]+):/gm;
  const relativeFirstCharacters = /* @__PURE__ */ new Set([".", "/"]);
  const isRelativeUrlWithoutProtocol = (relativeUrl) => relativeFirstCharacters.has(relativeUrl[0]);
  if (!url) {
    return "about:blank";
  }
  const sanitizedUrl = url.replaceAll(ctrlCharactersRegex, "").trim();
  if (isRelativeUrlWithoutProtocol(sanitizedUrl)) {
    return sanitizedUrl;
  }
  const urlSchemeParseResults = sanitizedUrl.match(urlSchemeRegex);
  if (!urlSchemeParseResults) {
    return sanitizedUrl;
  }
  const urlScheme = urlSchemeParseResults[0];
  if (invalidPrototcolRegex.test(urlScheme)) {
    return "about:blank";
  }
  return sanitizedUrl;
};
var cloneDeep = (obj) => JSON.parse(JSON.stringify(obj));
var isObject = (o6) => o6 && typeof o6 === "object";
var defaultsDeep = (target, defaults) => {
  Object.entries(defaults).forEach(([k2, v2]) => {
    if (k2 in target) {
      if (target[k2] === void 0) {
        target[k2] = v2;
      } else {
        target[k2] = isObject(v2) ? defaultsDeep(
          target[k2],
          v2
        ) : target[k2];
      }
    } else {
      target[k2] = v2;
    }
  });
  return target;
};
var uuidv4 = () => "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replaceAll(/[xy]/g, function(c4) {
  const r7 = Math.random() * 16 | 0;
  const v2 = c4 == "x" ? r7 : r7 & 3 | 8;
  return v2.toString(16);
});

// libs/embed-libs/shared-embed/src/lib/helpers/iframe-url.ts
var createIframeUrl = (dashboardContent) => {
  const searchString = [];
  const {
    appServer,
    language,
    qeVersion,
    loaderBackground,
    loaderSpinnerColor,
    loaderSpinnerBackground,
    loaderFontColor,
    screenMode,
    key,
    token,
    timezoneId,
    itemEmbedding,
    itemDimensions,
    dashboardId,
    dashboardSlug,
    itemId,
    mainColor,
    accentColor,
    editMode
  } = dashboardContent.params;
  if (language && language !== "auto") {
    searchString.push("language=" + language);
  }
  if (qeVersion) {
    searchString.push(`qe_version=${qeVersion}`);
  }
  if (loaderBackground) {
    searchString.push(`ldrbg=${loaderBackground}`);
  }
  if (loaderFontColor) {
    searchString.push(`ldrftclr=${loaderFontColor}`);
  }
  if (loaderSpinnerColor) {
    searchString.push(`ldrspclr=${loaderSpinnerColor}`);
  }
  if (loaderSpinnerBackground) {
    searchString.push(`ldrspbg=${loaderSpinnerBackground}`);
  }
  if (mainColor) {
    searchString.push(`mainColor=${mainColor}`);
  }
  if (accentColor) {
    searchString.push(`accentColor=${accentColor}`);
  }
  if (screenMode && screenMode !== "auto") {
    searchString.push("screenMode=" + screenMode);
  }
  if (timezoneId) {
    searchString.push("timezoneId=" + timezoneId);
  }
  if (itemEmbedding && itemDimensions && itemDimensions.width) {
    searchString.push("width=" + itemDimensions.width);
  }
  if (itemEmbedding && itemDimensions && itemDimensions.height) {
    searchString.push("height=" + itemDimensions.height);
  }
  if (editMode) {
    searchString.push("editMode=" + editMode);
  }
  const keyTokenUrlAnchor = key && token ? "#" + key + ":" + token : "";
  searchString.push("version=" + (dashboardContent.VERSION || VERSION));
  const iframeUrl = itemEmbedding ? `${appServer}i/${dashboardId || dashboardSlug}/${itemId}?${encodeURIComponent(
    searchString.join("&")
  )}${keyTokenUrlAnchor}` : `${appServer}i/${dashboardId || dashboardSlug}?${encodeURIComponent(
    searchString.join("&")
  )}${keyTokenUrlAnchor}`;
  return sanitizeUrl(iframeUrl);
};

// libs/embed-libs/shared-embed/src/lib/helpers/get-loader.ts
var getLoader = (dashboardComponent) => {
  let errorMsg;
  if (dashboardComponent.params.error || dashboardComponent.browserNotSupported.length > 0) {
    dashboardComponent.params.error = true;
    if (dashboardComponent.browserNotSupported) {
      errorMsg = "Browser is not supported.";
    }
  }
  const loadingMsg = translations_default.initializing[dashboardComponent.params.language] || translations_default.initializing.en;
  return dashboardComponent.browserNotSupported || dashboardComponent.params.error ? { loadingMsg, errorMsg } : { loadingMsg, errorMsg: "" };
};

// libs/embed-libs/shared-embed/src/lib/helpers/promise-dispatch.ts
function resolvePromise(dashboardComponent, requestId, data) {
  if (dashboardComponent.promises[requestId]) {
    dashboardComponent.promises[requestId].resolve(data);
    delete dashboardComponent.promises[requestId];
  }
}
function rejectPromise(dashboardComponent, requestId, data) {
  if (dashboardComponent.promises[requestId]) {
    dashboardComponent.promises[requestId].reject(data);
    delete dashboardComponent.promises[requestId];
  }
}

// libs/embed-libs/shared-embed/src/lib/helpers/styles.ts
function containerStyles(dashboardComponent) {
  const styles = {};
  if (dashboardComponent.params.loaderBackground && !dashboardComponent.iframeLoaded) {
    styles.background = dashboardComponent.params.loaderBackground;
  }
  return styles;
}
function containerLoaderStyles(dashboardComponent) {
  const height = dashboardComponent.params.itemDimensions?.height || DEFAULT_PARAMETERS.itemDimensions.height;
  const chartHeightDefined = dashboardComponent.params.itemEmbedding && height !== "auto";
  const display = dashboardComponent.iframeLoaded ? "none" : "";
  return {
    display,
    height: `${chartHeightDefined ? height : dashboardComponent.minHeight}px`
  };
}
function loaderStyles(dashboardComponent) {
  const background = dashboardComponent.params.loaderBackground || DEFAULT_PARAMETERS.loaderBackground;
  if (dashboardComponent.params.itemEmbedding) {
    dashboardComponent.loaderSize = 24;
    dashboardComponent.borderWidth = 4;
    dashboardComponent.minHeight = 200;
    const width = dashboardComponent.params.itemDimensions?.width || DEFAULT_PARAMETERS.itemDimensions.width;
    const height = dashboardComponent.params.itemDimensions?.height || DEFAULT_PARAMETERS.itemDimensions.height;
    const chartWidthDefined = width !== "auto";
    const chartHeightDefined = height !== "auto";
    return {
      background,
      width: chartWidthDefined ? `${width}px` : "100%",
      height: `${chartHeightDefined ? height : "200"}px`,
      "min-height": chartHeightDefined ? "" : "200px"
    };
  }
  return {
    background
  };
}
function plStyles(dashboardComponent) {
  const {
    loaderSpinnerColor: defaultSpinnerColor,
    loaderSpinnerBackground: defaultSpinnerBackground
  } = DEFAULT_PARAMETERS;
  const spinnerColorBorder = generateBorderStyle(
    dashboardComponent.borderWidth,
    dashboardComponent.params.loaderSpinnerColor || defaultSpinnerColor
  );
  const spinnerBackgroundBorder = generateBorderStyle(
    dashboardComponent.borderWidth,
    dashboardComponent.params.loaderSpinnerBackground || defaultSpinnerBackground
  );
  const size = `${dashboardComponent.loaderSize}px`;
  return {
    "border-top": spinnerColorBorder,
    "border-right": spinnerColorBorder,
    "border-bottom": spinnerBackgroundBorder,
    "border-left": spinnerBackgroundBorder,
    width: size,
    height: size
  };
}
function lcStyles(dashboardComponent) {
  const height = dashboardComponent.params.itemDimensions?.height || DEFAULT_PARAMETERS.itemDimensions.height;
  const chartHeightDefined = dashboardComponent.params.itemEmbedding && height !== "auto";
  const getPaddingStr = (value) => `${(value - dashboardComponent.loaderSize - 32) / 2}px`;
  const verticalPadding = dashboardComponent.params.itemId ? chartHeightDefined ? getPaddingStr(Number.parseInt(`${height}`, 10)) : getPaddingStr(dashboardComponent.minHeight) : "128px";
  return {
    padding: `${verticalPadding} 16px`
  };
}
function calculateIframeStyles(dashboardComponent) {
  const height = dashboardComponent.params.itemDimensions?.height || DEFAULT_PARAMETERS.itemDimensions.height;
  const width = dashboardComponent.params.itemDimensions?.width || DEFAULT_PARAMETERS.itemDimensions.width;
  if (dashboardComponent.params.itemEmbedding && height !== "auto") {
    dashboardComponent.iframeHeight = `${height}px`;
  } else if (dashboardComponent.params.itemEmbedding) {
    dashboardComponent.iframeHeight = "200px";
  } else if (["editFull", "editLimited"].includes(dashboardComponent.params.editMode)) {
    dashboardComponent.iframeHeight = "100%";
  } else {
    dashboardComponent.iframeHeight = "400px";
  }
  dashboardComponent.iframeWidth = dashboardComponent.params.itemEmbedding && width !== "auto" ? `${width}px` : "100%";
  const styles = {
    height: dashboardComponent.iframeHeight,
    width: dashboardComponent.iframeWidth
  };
  if (dashboardComponent.iframeLoaded) {
    styles.opacity = 1;
  } else if (dashboardComponent.error) {
    styles.opacity = 1;
    styles.width = "100%";
    styles.height = "400px";
  } else {
    styles.opacity = 0;
  }
  return styles;
}
function generateBorderStyle(borderWidth, color) {
  return `${borderWidth}px solid ${color}`;
}

// libs/embed-libs/shared-embed/src/lib/helpers/update-iframe-url.ts
var updateIframeUrl = (iframe, iframeUrl) => {
  if (iframe?.contentWindow) {
    iframe.contentWindow.location.replace(iframeUrl);
  }
};

// libs/embed-libs/shared-embed/src/lib/methods/export-dashboard.ts
var exportDashboard = (dashboardComponent, iframe, format) => {
  const requestId = uuidv4();
  if (iframe?.contentWindow) {
    iframe.contentWindow.postMessage(
      {
        action: "exportDashboard",
        format: format || "png",
        _version: dashboardComponent.VERSION,
        requestId
      },
      "*"
    );
    return new Promise((resolve, reject) => {
      dashboardComponent.promises[requestId] = { resolve, reject };
      setTimeout(() => {
        if (dashboardComponent.promises[requestId]) {
          const errorMessage = "exportDashboard request timed out.";
          dashboardComponent.promises[requestId].reject(
            new Error(errorMessage)
          );
          delete dashboardComponent.promises[requestId];
        }
      }, 12e5);
    });
  }
  return Promise.reject(new Error("Cannot retrieve contentWindow."));
};

// libs/embed-libs/shared-embed/src/lib/methods/get-accessible-dashboards.ts
var getAccessibleDashboards = (dashboardComponent) => {
  if ((dashboardComponent.currentAuthKey ?? dashboardComponent.authKey) && (dashboardComponent.currentAuthToken ?? dashboardComponent.authToken)) {
    const fetchUser = new Promise((resolve, reject) => {
      void (() => {
        try {
          const responsePromise = window.fetch(
            // With experimental=true URLs remove the trailing slash, since the web component expects them to not have it.
            // Fixes incorrect url with experimental=true
            `${addTrailingSlash(dashboardComponent.params.apiHost)}0.1.0/authorization`,
            {
              method: "POST",
              mode: "cors",
              cache: "no-cache",
              headers: { "Content-Type": "application/json" },
              redirect: "follow",
              body: JSON.stringify({
                action: "get",
                version: "0.1.0",
                key: dashboardComponent.currentAuthKey ?? dashboardComponent.authKey,
                token: dashboardComponent.currentAuthToken ?? dashboardComponent.authToken,
                find: {
                  where: { id: dashboardComponent.authKey },
                  attributes: ["id", "integration_id"],
                  include: [
                    {
                      model: "User",
                      attributes: ["id", "name", "locale_id"],
                      include: [{ model: "Group" }]
                    }
                  ]
                }
              })
            }
          );
          responsePromise.then((response) => response.json()).then((data) => {
            if (!data || !data.rows || data.rows.length === 0 || !data.rows[0].user) {
              reject("Authorization failed, please check authKey and authToken. If the problem persists, please contact support@luzmo.com.");
            } else {
              data.rows[0].user.userGroups = data.rows[0].user.groups.map(
                (group) => group.id
              );
              delete data.rows[0].user.groups;
              resolve({
                integrationId: data.rows[0].integration_id,
                user: data.rows[0].user
              });
            }
          });
        } catch (error) {
          reject(error);
        }
      })();
    });
    return new Promise((resolve, reject) => void (() => {
      try {
        fetchUser.then((authorizationData) => {
          const responsePromise = window.fetch(
            // With experimental=true URLs remove the trailing slash, since the web component expects them to not have it.
            // Fixes incorrect url with experimental=true
            `${addTrailingSlash(dashboardComponent.params.apiHost)}0.1.0/securable`,
            {
              method: "POST",
              mode: "cors",
              cache: "no-cache",
              headers: { "Content-Type": "application/json" },
              redirect: "follow",
              body: JSON.stringify({
                action: "get",
                version: "0.1.0",
                key: dashboardComponent.currentAuthKey ?? dashboardComponent.authKey,
                token: dashboardComponent.currentAuthToken ?? dashboardComponent.authToken,
                find: {
                  where: { type: "dashboard", is_variant: false },
                  attributes: ["id", "name", "modified_at"],
                  order: [["modified_at", "desc"]],
                  include: [
                    { model: "Tag", attributes: ["tag"] },
                    { model: "User", attributes: ["id", "name"] },
                    {
                      model: "Group",
                      attributes: ["id", "name", "public"]
                    },
                    { model: "Integration", attributes: ["id", "name"] }
                  ]
                }
              })
            }
          );
          responsePromise.then((response) => response.json()).then((data) => {
            const dashboardList = data.rows.length > 0 ? formatDashboardList(
              data.rows,
              authorizationData.integrationId,
              authorizationData.user,
              dashboardComponent
            ) : [];
            resolve(dashboardList);
          });
        }).catch((error) => {
          reject(error);
        });
      } catch (error) {
        reject(error);
      }
    })());
  }
  throw new Error("No authKey or authToken found in the luzmo-dashboard component.");
};

// libs/embed-libs/shared-embed/src/lib/methods/get-data.ts
var getData = (dashboardComponent, iframe, itemId) => {
  if (iframe?.contentWindow) {
    const requestId = uuidv4();
    iframe.contentWindow.postMessage(
      {
        action: "getData",
        id: itemId,
        _version: dashboardComponent.VERSION,
        requestId
      },
      "*"
    );
    return new Promise((resolve, reject) => {
      dashboardComponent.promises[requestId] = { resolve, reject };
      setTimeout(() => {
        if (dashboardComponent.promises[requestId]) {
          const errorMessage = "getData request timed out.";
          dashboardComponent.promises[requestId].reject(
            new Error(errorMessage)
          );
          delete dashboardComponent.promises[requestId];
        }
      }, 2e3);
    });
  }
  return Promise.reject(new Error("Cannot retrieve contentWindow."));
};

// libs/embed-libs/shared-embed/src/lib/methods/get-filters.ts
var getFilters = (dashboardComponent, iframe) => {
  if (iframe?.contentWindow) {
    const requestId = uuidv4();
    iframe.contentWindow.postMessage(
      { action: "getFilters", _version: dashboardComponent.VERSION, requestId },
      "*"
    );
    return new Promise((resolve, reject) => {
      dashboardComponent.promises[requestId] = { resolve, reject };
      setTimeout(() => {
        if (dashboardComponent.promises[requestId]) {
          const errorMessage = "getFilters request timed out.";
          dashboardComponent.promises[requestId].reject(
            new Error(errorMessage)
          );
          delete dashboardComponent.promises[requestId];
        }
      }, 2e3);
    });
  }
  return Promise.reject(new Error("Cannot retrieve contentWindow."));
};

// libs/embed-libs/shared-embed/src/lib/methods/refresh-data.ts
var refreshData = (dashboardComponent, iframe, itemId) => {
  if (iframe) {
    iframe.contentWindow.postMessage(
      {
        action: "refreshData",
        id: itemId ?? null,
        _version: dashboardComponent.VERSION
      },
      "*"
    );
    return Promise.resolve();
  }
  return Promise.reject(new Error("Cannot retrieve contentWindow."));
};

// libs/embed-libs/shared-embed/src/lib/methods/reload-dashboard.ts
var reloadDashboard = (dashboardComponent, iframe) => {
  if (iframe) {
    iframe.contentWindow.postMessage(
      { action: "reloadDashboard", _version: dashboardComponent.VERSION },
      "*"
    );
    return Promise.resolve();
  }
  return Promise.reject(new Error("Cannot retrieve contentWindow."));
};

// libs/embed-libs/shared-embed/src/lib/methods/set-authorization.ts
var setAuthorization = (dashboardComponent, iframe, key, token) => {
  if (iframe) {
    iframe.contentWindow.postMessage(
      {
        action: "setAuthorization",
        key,
        token,
        _version: dashboardComponent.VERSION
      },
      "*"
    );
    return Promise.resolve();
  }
  return Promise.reject(new Error("Cannot retrieve contentWindow."));
};

// libs/embed-libs/shared-embed/src/lib/methods/set-edit-mode.ts
var setEditMode = (dashboardComponent, iframe, editMode) => {
  if (iframe && iframe.contentWindow) {
    const requestId = uuidv4();
    if (dashboardComponent.params.editMode === editMode) {
      return Promise.reject(
        new Error("editMode parameter is currently already set to " + editMode)
      );
    } else if (editMode !== "editFull" && editMode !== "editLimited" && editMode !== "view") {
      return Promise.reject(
        new Error(
          "The editMode parameter must be one of the following values: 'editFull', 'editLimited' or 'view'."
        )
      );
    }
    iframe.contentWindow.postMessage(
      {
        action: "setEditMode",
        editMode,
        _version: dashboardComponent.VERSION,
        requestId
      },
      "*"
    );
    return new Promise((resolve, reject) => {
      dashboardComponent.promises[requestId] = { resolve, reject };
    });
  }
  return Promise.reject(new Error("Cannot retrieve contentWindow."));
};

// libs/embed-libs/shared-embed/src/lib/methods/set-filters.ts
var setFilters = (dashboardComponent, iframe, itemId, filters) => {
  if (iframe) {
    const requestId = uuidv4();
    iframe.contentWindow.postMessage(
      {
        action: "setFilters",
        itemId,
        filters,
        _version: dashboardComponent.VERSION,
        requestId
      },
      "*"
    );
    return new Promise((resolve, reject) => {
      dashboardComponent.promises[requestId] = { resolve, reject };
      setTimeout(() => {
        if (dashboardComponent.promises[requestId]) {
          const errorMessage = "setFilters request timed out.";
          dashboardComponent.promises[requestId].reject(
            new Error(errorMessage)
          );
          delete dashboardComponent.promises[requestId];
        }
      }, 2e3);
    });
  }
  return Promise.reject();
};

// libs/embed-libs/shared-embed/src/lib/methods/set-preview.ts
var setPreview = (dashboardComponent, iframe, preview) => {
  if (iframe) {
    const requestId = uuidv4();
    iframe.contentWindow.postMessage(
      {
        action: "setPreview",
        preview,
        requestId,
        _version: dashboardComponent.VERSION
      },
      "*"
    );
    return new Promise((resolve, reject) => {
      dashboardComponent.promises[requestId] = { resolve, reject };
      setTimeout(() => {
        if (dashboardComponent.promises[requestId]) {
          const errorMessage = "setPreview request timed out.";
          dashboardComponent.promises[requestId].reject(
            new Error(errorMessage)
          );
          delete dashboardComponent.promises[requestId];
        }
      }, 2e3);
    });
  }
  return Promise.reject(new Error("Cannot retrieve contentWindow."));
};

// libs/embed-libs/shared-embed/src/lib/service/calculate-dimensions.ts
var _setCalculatedDimensions = (dashboardComponent) => {
  if (!dashboardComponent.dimensions) {
    dashboardComponent.params.error = true;
    _handleError("Got no valid dimensions.", dashboardComponent.params);
    return;
  }
  const dashboardDimensionKeys = Object.keys(
    dashboardComponent.dimensions
  );
  const availableScreenModeExtents = [];
  let currentMode;
  let currentWidth = 0;
  let currentHeight = 0;
  if (!dashboardComponent.params.switchScreenModeOnResize && dashboardComponent.currentScreenMode) {
    dashboardComponent.params.screenMode = dashboardComponent.currentScreenMode;
  }
  if (dashboardComponent.dimensions && dashboardDimensionKeys?.[0] === "fixed") {
    currentMode = "fixed";
    currentWidth = dashboardComponent.dimensions.fixed[0];
    currentHeight = dashboardComponent.dimensions.fixed[1];
  } else {
    const screenModes = [
      "mobile",
      "tablet",
      "desktop",
      "largeScreen"
    ];
    if (dashboardComponent.params.itemEmbedding) {
      const [firstMatch] = dashboardDimensionKeys;
      let match = cloneDeep(firstMatch);
      for (const screenMode of screenModes) {
        const found = dashboardDimensionKeys.find((key) => key === screenMode);
        if (found) {
          match = cloneDeep(found);
        } else {
          dashboardComponent.dimensions[screenMode] = dashboardComponent.dimensions[match];
        }
      }
    }
    for (const key of dashboardDimensionKeys) {
      if (screenModes.includes(key)) {
        const screenMode = EXTENTS.find((extent) => extent.mode === key);
        availableScreenModeExtents.push({
          ...screenMode,
          height: dashboardComponent.dimensions[key][1] || 50
        });
      }
    }
    if (!availableScreenModeExtents.map((screenModeExtent) => screenModeExtent.mode).includes(dashboardComponent.params.screenMode)) {
      dashboardComponent.params.screenMode = "auto";
    }
    if (dashboardComponent.params.screenMode === "auto") {
      availableScreenModeExtents.sort(
        (a3, b3) => screenModes.indexOf(a3.mode) - screenModes.indexOf(b3.mode)
      );
      for (const screenModeExtent of availableScreenModeExtents) {
        const [screenModeExtentMinWidth, screenModeExtentMaxWidth] = screenModeExtent.width;
        if (dashboardComponent.params.itemEmbedding && dashboardComponent.params.itemDimensions && dashboardComponent.params.itemDimensions.width !== "auto" && dashboardComponent.params.itemDimensions.height === "auto") {
          if (screenModeExtentMinWidth < dashboardComponent.params.itemDimensions.width) {
            currentMode = screenModeExtent.mode;
            currentWidth = Number.parseInt(
              dashboardComponent.params.itemDimensions.width,
              10
            );
            currentHeight = dashboardComponent.dimensions[currentMode][1];
          }
        } else {
          if (screenModeExtentMinWidth < dashboardComponent.containerWidth) {
            currentMode = screenModeExtent.mode;
            if (dashboardComponent.params.itemEmbedding) {
              currentWidth = dashboardComponent.params.itemDimensions?.width && dashboardComponent.params.itemDimensions?.width !== "auto" ? Number.parseInt(
                dashboardComponent.params.itemDimensions.width,
                10
              ) : Math.min(
                dashboardComponent.containerWidth,
                screenModeExtentMaxWidth
              );
              currentHeight = dashboardComponent.params.itemDimensions?.height && dashboardComponent.params.itemDimensions?.height !== "auto" ? Number.parseInt(
                dashboardComponent.params.itemDimensions.height,
                10
              ) : dashboardComponent.dimensions[currentMode][1];
              if (dashboardComponent.params.itemDimensions && dashboardComponent.params.itemDimensions.width !== "auto" && dashboardComponent.params.itemDimensions.height === "auto") {
                currentWidth = Number.parseInt(
                  dashboardComponent.params.itemDimensions.width,
                  10
                );
                currentHeight = dashboardComponent.dimensions[currentMode][1];
              }
            } else {
              currentWidth = Math.max(
                Math.min(
                  screenModeExtentMaxWidth,
                  dashboardComponent.containerWidth
                ),
                screenModeExtentMinWidth
              );
              currentHeight = screenModeExtent.height;
            }
          }
        }
      }
    } else {
      const {
        height: chosenModeHeight,
        width: [chosenModeMinWidth, chosenModeMaxWidth]
      } = availableScreenModeExtents.find(
        (d3) => d3.mode === dashboardComponent.params.screenMode
      );
      currentMode = dashboardComponent.params.screenMode;
      currentWidth = Math.max(
        Math.min(chosenModeMaxWidth, dashboardComponent.containerWidth),
        chosenModeMinWidth
      );
      currentHeight = chosenModeHeight;
      if (dashboardComponent.params.itemEmbedding) {
        if (dashboardComponent.params.itemDimensions?.width && dashboardComponent.params.itemDimensions?.width !== "auto") {
          currentWidth = Number.parseInt(
            dashboardComponent.params.itemDimensions.width,
            10
          );
        }
        if (dashboardComponent.params.itemDimensions?.height && dashboardComponent.params.itemDimensions?.height !== "auto") {
          currentHeight = Number.parseInt(
            dashboardComponent.params.itemDimensions.height,
            10
          );
        }
      }
    }
    if (!currentMode && availableScreenModeExtents.length > 0) {
      const [lowestModeExtent] = availableScreenModeExtents;
      currentMode = lowestModeExtent.mode;
      currentWidth = lowestModeExtent.width[0];
      currentHeight = lowestModeExtent.height;
    }
  }
  dashboardComponent.currentScreenMode = currentMode;
  const style = {
    height: currentHeight === 0 ? "100%" : `${String(currentHeight)}px`,
    width: currentWidth === 0 ? "100%" : `${String(currentWidth)}px`,
    opacity: 1
  };
  const dashboardEditModes = ["editFull", "editLimited"];
  if (dashboardEditModes.includes(dashboardComponent.params.editMode)) {
    style.height = "100%";
    style.width = "100%";
  } else if (dashboardComponent.params.itemEmbedding && dashboardComponent.params.itemDimensions) {
    style.height = dashboardComponent.params.itemDimensions.height === "auto" ? style.height : `${String(dashboardComponent.params.itemDimensions.height).replace(
      "px",
      ""
    )}px`;
    style.width = dashboardComponent.params.itemDimensions.width === "auto" ? style.width : `${String(dashboardComponent.params.itemDimensions.width).replace(
      "px",
      ""
    )}px`;
  } else {
    style.height = currentHeight === 0 ? "100%" : `${String(currentHeight)}px`;
    style.width = currentWidth === 0 ? "100%" : `${String(currentWidth)}px`;
  }
  dashboardComponent.currentScreenMode = currentMode;
  dashboardComponent.iframeStyle = style;
};

// libs/embed-libs/shared-embed/src/lib/service/extend-and-validate.ts
var _extendAndValidate = (dashboardContent) => {
  const defaultKeys = Object.keys(
    DEFAULT_PARAMETERS
  );
  const errorMessages = [];
  const requiredStringParams = [
    "appServer",
    "language",
    "apiHost"
  ];
  const requiredStringOrEmptyParams = [
    "key",
    "token",
    "screenMode",
    "timezoneId",
    "itemId",
    "container",
    "loaderBackground",
    "loaderFontColor",
    "loaderSpinnerColor",
    "loaderSpinnerBackground"
  ];
  const objectParams = ["itemDimensions"];
  const screenModes = [
    "mobile",
    "tablet",
    "desktop",
    "largeScreen",
    "fixed",
    "auto"
  ];
  for (const key of defaultKeys) {
    if (dashboardContent.params[key] === null) {
      delete dashboardContent.params[key];
    }
  }
  dashboardContent.params = defaultsDeep(
    dashboardContent.params,
    DEFAULT_PARAMETERS
  );
  if ((!dashboardContent.params.dashboardId || typeof dashboardContent.params.dashboardId !== "string") && (!dashboardContent.params.dashboardSlug || typeof dashboardContent.params.dashboardSlug !== "string")) {
    delete dashboardContent.params.dashboardId;
    delete dashboardContent.params.dashboardSlug;
  }
  for (const key of requiredStringParams) {
    if (typeof dashboardContent.params[key] !== "string") {
      delete dashboardContent.params[key];
      errorMessages.push(`${key} needs to be of type string.`);
    }
  }
  for (const key of requiredStringOrEmptyParams) {
    if (typeof dashboardContent.params[key] !== "string" && dashboardContent.params[key] !== null && dashboardContent.params[key] !== void 0) {
      delete dashboardContent.params[key];
      errorMessages.push(`${key} needs to be of type string or empty.`);
    }
  }
  if (dashboardContent.params.screenMode && !screenModes.includes(dashboardContent.params.screenMode)) {
    const validTypesString = screenModes.map((mode) => `"${mode}"`).join(", ");
    errorMessages.push(
      `"screenMode" needs should be one of these types: ${validTypesString}`
    );
    dashboardContent.params.screenMode = "auto";
  }
  if (dashboardContent.params.itemId) {
    dashboardContent.params.itemEmbedding = true;
  }
  for (const key of objectParams) {
    if (typeof dashboardContent.params[key] !== "object" && dashboardContent.params[key] !== null) {
      delete dashboardContent.params[key];
      errorMessages.push(`${key} needs to be of type object or empty.`);
    }
  }
  if (typeof dashboardContent.params.switchScreenModeOnResize !== "boolean") {
    errorMessages.push("switchScreenModeOnResize needs to be of type boolean.");
    dashboardContent.params.switchScreenModeOnResize = true;
  }
  if (dashboardContent.params.itemDimensions) {
    const itemDimensionsKeys = [
      "width",
      "height"
    ];
    for (const key of itemDimensionsKeys) {
      if (!["string", "number"].includes(
        typeof dashboardContent.params.itemDimensions[key]
      ) && dashboardContent.params.itemDimensions[key] !== null) {
        errorMessages.push(
          `itemDimensions ${key} needs to be of type string, number or empty.`
        );
      }
    }
  }
  dashboardContent.params = defaultsDeep(
    dashboardContent.params,
    DEFAULT_PARAMETERS
  );
  if (dashboardContent.params.screenMode && dashboardContent.params.screenMode !== "auto") {
    dashboardContent.params.switchScreenModeOnResize = false;
  }
  if (dashboardContent.params.itemDimensions.width !== "auto") {
    dashboardContent.params.itemDimensions.width = Number.parseInt(
      `${dashboardContent.params.itemDimensions.width}`.replace("px", ""),
      10
    ) || "auto";
  }
  if (dashboardContent.params.itemDimensions.height !== "auto") {
    dashboardContent.params.itemDimensions.height = Number.parseInt(
      `${dashboardContent.params.itemDimensions.height}`.replace("px", ""),
      10
    ) || "auto";
  }
  dashboardContent.params.itemEmbedding = !!dashboardContent.params.itemId;
  if (dashboardContent.params.error || errorMessages.length > 0) {
    _handleError(errorMessages.join("\n"), dashboardContent.params);
  }
};

// libs/embed-libs/shared-embed/src/lib/constants/class-constants.ts
var luzmoEmbedContainerClass = "luzmo-embed-container";

// libs/embed-libs/shared-embed/src/lib/helpers/remote-federation.ts
var remotesMap = {
  // "remote_app": { url: "http://localhost:4201/remoteEntry.mjs", format: "esm", from: "vite" },
};
var loadJS = async (url, fn) => {
  const resolvedUrl = typeof url === "function" ? await url() : url;
  const script = document.createElement("script");
  script.type = "text/javascript";
  script.addEventListener("load", fn);
  script.src = resolvedUrl;
  document.querySelectorAll("head")[0].append(script);
};
var wrapShareModule = (remoteFrom) => ({});
async function __federationMethodEnsure(remoteId) {
  const remote = remotesMap[remoteId];
  if (remote.inited) {
    return remote.lib;
  }
  if (remote.format === "var") {
    return new Promise((resolve) => {
      const callback = () => {
        if (!remote.inited) {
          remote.lib = window[remoteId];
          remote.lib.init(wrapShareModule(remote.from));
          remote.inited = true;
        }
        resolve(remote.lib);
      };
      return loadJS(remote.url, callback);
    });
  } else if (["esm", "systemjs"].includes(remote.format)) {
    return new Promise((resolve, reject) => {
      const getUrl = typeof remote.url === "function" ? remote.url : () => Promise.resolve(remote.url);
      getUrl().then((url) => {
        import(
          /* @vite-ignore */
          /* webpackIgnore: true */
          url
        ).then((lib) => {
          if (!remote.inited) {
            const shareScope = wrapShareModule(remote.from);
            lib.init(shareScope);
            remote.lib = lib;
            remote.lib.init(shareScope);
            remote.inited = true;
          }
          resolve(remote.lib);
        }).catch(reject);
      });
    });
  }
}
function __federationMethodWrapDefault(module2, need) {
  if (!(module2 == null ? void 0 : module2.default) && need) {
    const obj = /* @__PURE__ */ Object.create(null);
    obj.default = module2;
    obj.__esModule = true;
    return obj;
  }
  return module2;
}
async function __federationMethodGetRemote(remoteName, componentName) {
  return __federationMethodEnsure(remoteName).then((remote) => remote.get(componentName).then((factory) => factory()));
}

// node_modules/lit-html/lit-html.js
var t3 = globalThis;
var i3 = t3.trustedTypes;
var s2 = i3 ? i3.createPolicy("lit-html", { createHTML: (t5) => t5 }) : void 0;
var e5 = "$lit$";
var h2 = `lit$${(Math.random() + "").slice(9)}$`;
var o4 = "?" + h2;
var n4 = `<${o4}>`;
var r5 = document;
var l2 = () => r5.createComment("");
var c3 = (t5) => null === t5 || "object" != typeof t5 && "function" != typeof t5;
var a2 = Array.isArray;
var u2 = (t5) => a2(t5) || "function" == typeof t5?.[Symbol.iterator];
var d2 = "[ 	\n\f\r]";
var f2 = /<(?:(!--|\/[^a-zA-Z])|(\/?[a-zA-Z][^>\s]*)|(\/?$))/g;
var v = /-->/g;
var _ = />/g;
var m = RegExp(`>|${d2}(?:([^\\s"'>=/]+)(${d2}*=${d2}*(?:[^ 	
\f\r"'\`<>=]|("|')|))|$)`, "g");
var p2 = /'/g;
var g = /"/g;
var $ = /^(?:script|style|textarea|title)$/i;
var y2 = (t5) => (i6, ...s4) => ({ _$litType$: t5, strings: i6, values: s4 });
var x = y2(1);
var b2 = y2(2);
var w = Symbol.for("lit-noChange");
var T = Symbol.for("lit-nothing");
var A = /* @__PURE__ */ new WeakMap();
var E = r5.createTreeWalker(r5, 129);
function C(t5, i6) {
  if (!Array.isArray(t5) || !t5.hasOwnProperty("raw"))
    throw Error("invalid template strings array");
  return void 0 !== s2 ? s2.createHTML(i6) : i6;
}
var P = (t5, i6) => {
  const s4 = t5.length - 1, o6 = [];
  let r7, l3 = 2 === i6 ? "<svg>" : "", c4 = f2;
  for (let i7 = 0; i7 < s4; i7++) {
    const s5 = t5[i7];
    let a3, u3, d3 = -1, y3 = 0;
    for (; y3 < s5.length && (c4.lastIndex = y3, u3 = c4.exec(s5), null !== u3); )
      y3 = c4.lastIndex, c4 === f2 ? "!--" === u3[1] ? c4 = v : void 0 !== u3[1] ? c4 = _ : void 0 !== u3[2] ? ($.test(u3[2]) && (r7 = RegExp("</" + u3[2], "g")), c4 = m) : void 0 !== u3[3] && (c4 = m) : c4 === m ? ">" === u3[0] ? (c4 = r7 ?? f2, d3 = -1) : void 0 === u3[1] ? d3 = -2 : (d3 = c4.lastIndex - u3[2].length, a3 = u3[1], c4 = void 0 === u3[3] ? m : '"' === u3[3] ? g : p2) : c4 === g || c4 === p2 ? c4 = m : c4 === v || c4 === _ ? c4 = f2 : (c4 = m, r7 = void 0);
    const x2 = c4 === m && t5[i7 + 1].startsWith("/>") ? " " : "";
    l3 += c4 === f2 ? s5 + n4 : d3 >= 0 ? (o6.push(a3), s5.slice(0, d3) + e5 + s5.slice(d3) + h2 + x2) : s5 + h2 + (-2 === d3 ? i7 : x2);
  }
  return [C(t5, l3 + (t5[s4] || "<?>") + (2 === i6 ? "</svg>" : "")), o6];
};
var V = class _V {
  constructor({ strings: t5, _$litType$: s4 }, n6) {
    let r7;
    this.parts = [];
    let c4 = 0, a3 = 0;
    const u3 = t5.length - 1, d3 = this.parts, [f3, v2] = P(t5, s4);
    if (this.el = _V.createElement(f3, n6), E.currentNode = this.el.content, 2 === s4) {
      const t6 = this.el.content.firstChild;
      t6.replaceWith(...t6.childNodes);
    }
    for (; null !== (r7 = E.nextNode()) && d3.length < u3; ) {
      if (1 === r7.nodeType) {
        if (r7.hasAttributes())
          for (const t6 of r7.getAttributeNames())
            if (t6.endsWith(e5)) {
              const i6 = v2[a3++], s5 = r7.getAttribute(t6).split(h2), e7 = /([.?@])?(.*)/.exec(i6);
              d3.push({ type: 1, index: c4, name: e7[2], strings: s5, ctor: "." === e7[1] ? k : "?" === e7[1] ? H : "@" === e7[1] ? I : R }), r7.removeAttribute(t6);
            } else
              t6.startsWith(h2) && (d3.push({ type: 6, index: c4 }), r7.removeAttribute(t6));
        if ($.test(r7.tagName)) {
          const t6 = r7.textContent.split(h2), s5 = t6.length - 1;
          if (s5 > 0) {
            r7.textContent = i3 ? i3.emptyScript : "";
            for (let i6 = 0; i6 < s5; i6++)
              r7.append(t6[i6], l2()), E.nextNode(), d3.push({ type: 2, index: ++c4 });
            r7.append(t6[s5], l2());
          }
        }
      } else if (8 === r7.nodeType)
        if (r7.data === o4)
          d3.push({ type: 2, index: c4 });
        else {
          let t6 = -1;
          for (; -1 !== (t6 = r7.data.indexOf(h2, t6 + 1)); )
            d3.push({ type: 7, index: c4 }), t6 += h2.length - 1;
        }
      c4++;
    }
  }
  static createElement(t5, i6) {
    const s4 = r5.createElement("template");
    return s4.innerHTML = t5, s4;
  }
};
function N(t5, i6, s4 = t5, e7) {
  if (i6 === w)
    return i6;
  let h3 = void 0 !== e7 ? s4._$Co?.[e7] : s4._$Cl;
  const o6 = c3(i6) ? void 0 : i6._$litDirective$;
  return h3?.constructor !== o6 && (h3?._$AO?.(false), void 0 === o6 ? h3 = void 0 : (h3 = new o6(t5), h3._$AT(t5, s4, e7)), void 0 !== e7 ? (s4._$Co ??= [])[e7] = h3 : s4._$Cl = h3), void 0 !== h3 && (i6 = N(t5, h3._$AS(t5, i6.values), h3, e7)), i6;
}
var S2 = class {
  constructor(t5, i6) {
    this._$AV = [], this._$AN = void 0, this._$AD = t5, this._$AM = i6;
  }
  get parentNode() {
    return this._$AM.parentNode;
  }
  get _$AU() {
    return this._$AM._$AU;
  }
  u(t5) {
    const { el: { content: i6 }, parts: s4 } = this._$AD, e7 = (t5?.creationScope ?? r5).importNode(i6, true);
    E.currentNode = e7;
    let h3 = E.nextNode(), o6 = 0, n6 = 0, l3 = s4[0];
    for (; void 0 !== l3; ) {
      if (o6 === l3.index) {
        let i7;
        2 === l3.type ? i7 = new M(h3, h3.nextSibling, this, t5) : 1 === l3.type ? i7 = new l3.ctor(h3, l3.name, l3.strings, this, t5) : 6 === l3.type && (i7 = new L(h3, this, t5)), this._$AV.push(i7), l3 = s4[++n6];
      }
      o6 !== l3?.index && (h3 = E.nextNode(), o6++);
    }
    return E.currentNode = r5, e7;
  }
  p(t5) {
    let i6 = 0;
    for (const s4 of this._$AV)
      void 0 !== s4 && (void 0 !== s4.strings ? (s4._$AI(t5, s4, i6), i6 += s4.strings.length - 2) : s4._$AI(t5[i6])), i6++;
  }
};
var M = class _M {
  get _$AU() {
    return this._$AM?._$AU ?? this._$Cv;
  }
  constructor(t5, i6, s4, e7) {
    this.type = 2, this._$AH = T, this._$AN = void 0, this._$AA = t5, this._$AB = i6, this._$AM = s4, this.options = e7, this._$Cv = e7?.isConnected ?? true;
  }
  get parentNode() {
    let t5 = this._$AA.parentNode;
    const i6 = this._$AM;
    return void 0 !== i6 && 11 === t5?.nodeType && (t5 = i6.parentNode), t5;
  }
  get startNode() {
    return this._$AA;
  }
  get endNode() {
    return this._$AB;
  }
  _$AI(t5, i6 = this) {
    t5 = N(this, t5, i6), c3(t5) ? t5 === T || null == t5 || "" === t5 ? (this._$AH !== T && this._$AR(), this._$AH = T) : t5 !== this._$AH && t5 !== w && this._(t5) : void 0 !== t5._$litType$ ? this.$(t5) : void 0 !== t5.nodeType ? this.T(t5) : u2(t5) ? this.k(t5) : this._(t5);
  }
  S(t5) {
    return this._$AA.parentNode.insertBefore(t5, this._$AB);
  }
  T(t5) {
    this._$AH !== t5 && (this._$AR(), this._$AH = this.S(t5));
  }
  _(t5) {
    this._$AH !== T && c3(this._$AH) ? this._$AA.nextSibling.data = t5 : this.T(r5.createTextNode(t5)), this._$AH = t5;
  }
  $(t5) {
    const { values: i6, _$litType$: s4 } = t5, e7 = "number" == typeof s4 ? this._$AC(t5) : (void 0 === s4.el && (s4.el = V.createElement(C(s4.h, s4.h[0]), this.options)), s4);
    if (this._$AH?._$AD === e7)
      this._$AH.p(i6);
    else {
      const t6 = new S2(e7, this), s5 = t6.u(this.options);
      t6.p(i6), this.T(s5), this._$AH = t6;
    }
  }
  _$AC(t5) {
    let i6 = A.get(t5.strings);
    return void 0 === i6 && A.set(t5.strings, i6 = new V(t5)), i6;
  }
  k(t5) {
    a2(this._$AH) || (this._$AH = [], this._$AR());
    const i6 = this._$AH;
    let s4, e7 = 0;
    for (const h3 of t5)
      e7 === i6.length ? i6.push(s4 = new _M(this.S(l2()), this.S(l2()), this, this.options)) : s4 = i6[e7], s4._$AI(h3), e7++;
    e7 < i6.length && (this._$AR(s4 && s4._$AB.nextSibling, e7), i6.length = e7);
  }
  _$AR(t5 = this._$AA.nextSibling, i6) {
    for (this._$AP?.(false, true, i6); t5 && t5 !== this._$AB; ) {
      const i7 = t5.nextSibling;
      t5.remove(), t5 = i7;
    }
  }
  setConnected(t5) {
    void 0 === this._$AM && (this._$Cv = t5, this._$AP?.(t5));
  }
};
var R = class {
  get tagName() {
    return this.element.tagName;
  }
  get _$AU() {
    return this._$AM._$AU;
  }
  constructor(t5, i6, s4, e7, h3) {
    this.type = 1, this._$AH = T, this._$AN = void 0, this.element = t5, this.name = i6, this._$AM = e7, this.options = h3, s4.length > 2 || "" !== s4[0] || "" !== s4[1] ? (this._$AH = Array(s4.length - 1).fill(new String()), this.strings = s4) : this._$AH = T;
  }
  _$AI(t5, i6 = this, s4, e7) {
    const h3 = this.strings;
    let o6 = false;
    if (void 0 === h3)
      t5 = N(this, t5, i6, 0), o6 = !c3(t5) || t5 !== this._$AH && t5 !== w, o6 && (this._$AH = t5);
    else {
      const e8 = t5;
      let n6, r7;
      for (t5 = h3[0], n6 = 0; n6 < h3.length - 1; n6++)
        r7 = N(this, e8[s4 + n6], i6, n6), r7 === w && (r7 = this._$AH[n6]), o6 ||= !c3(r7) || r7 !== this._$AH[n6], r7 === T ? t5 = T : t5 !== T && (t5 += (r7 ?? "") + h3[n6 + 1]), this._$AH[n6] = r7;
    }
    o6 && !e7 && this.j(t5);
  }
  j(t5) {
    t5 === T ? this.element.removeAttribute(this.name) : this.element.setAttribute(this.name, t5 ?? "");
  }
};
var k = class extends R {
  constructor() {
    super(...arguments), this.type = 3;
  }
  j(t5) {
    this.element[this.name] = t5 === T ? void 0 : t5;
  }
};
var H = class extends R {
  constructor() {
    super(...arguments), this.type = 4;
  }
  j(t5) {
    this.element.toggleAttribute(this.name, !!t5 && t5 !== T);
  }
};
var I = class extends R {
  constructor(t5, i6, s4, e7, h3) {
    super(t5, i6, s4, e7, h3), this.type = 5;
  }
  _$AI(t5, i6 = this) {
    if ((t5 = N(this, t5, i6, 0) ?? T) === w)
      return;
    const s4 = this._$AH, e7 = t5 === T && s4 !== T || t5.capture !== s4.capture || t5.once !== s4.once || t5.passive !== s4.passive, h3 = t5 !== T && (s4 === T || e7);
    e7 && this.element.removeEventListener(this.name, this, s4), h3 && this.element.addEventListener(this.name, this, t5), this._$AH = t5;
  }
  handleEvent(t5) {
    "function" == typeof this._$AH ? this._$AH.call(this.options?.host ?? this.element, t5) : this._$AH.handleEvent(t5);
  }
};
var L = class {
  constructor(t5, i6, s4) {
    this.element = t5, this.type = 6, this._$AN = void 0, this._$AM = i6, this.options = s4;
  }
  get _$AU() {
    return this._$AM._$AU;
  }
  _$AI(t5) {
    N(this, t5);
  }
};
var Z = t3.litHtmlPolyfillSupport;
Z?.(V, M), (t3.litHtmlVersions ??= []).push("3.1.2");
var j = (t5, i6, s4) => {
  const e7 = s4?.renderBefore ?? i6;
  let h3 = e7._$litPart$;
  if (void 0 === h3) {
    const t6 = s4?.renderBefore ?? null;
    e7._$litPart$ = h3 = new M(i6.insertBefore(l2(), t6), t6, void 0, s4 ?? {});
  }
  return h3._$AI(t5), h3;
};

// node_modules/lit-element/lit-element.js
var s3 = class extends b {
  constructor() {
    super(...arguments), this.renderOptions = { host: this }, this._$Do = void 0;
  }
  createRenderRoot() {
    const t5 = super.createRenderRoot();
    return this.renderOptions.renderBefore ??= t5.firstChild, t5;
  }
  update(t5) {
    const i6 = this.render();
    this.hasUpdated || (this.renderOptions.isConnected = this.isConnected), super.update(t5), this._$Do = j(i6, this.renderRoot, this.renderOptions);
  }
  connectedCallback() {
    super.connectedCallback(), this._$Do?.setConnected(true);
  }
  disconnectedCallback() {
    super.disconnectedCallback(), this._$Do?.setConnected(false);
  }
  render() {
    return w;
  }
};
s3._$litElement$ = true, s3["finalized", "finalized"] = true, globalThis.litElementHydrateSupport?.({ LitElement: s3 });
var r6 = globalThis.litElementPolyfillSupport;
r6?.({ LitElement: s3 });
(globalThis.litElementVersions ??= []).push("4.0.4");

// node_modules/lit-html/directive.js
var t4 = { ATTRIBUTE: 1, CHILD: 2, PROPERTY: 3, BOOLEAN_ATTRIBUTE: 4, EVENT: 5, ELEMENT: 6 };
var e6 = (t5) => (...e7) => ({ _$litDirective$: t5, values: e7 });
var i4 = class {
  constructor(t5) {
  }
  get _$AU() {
    return this._$AM._$AU;
  }
  _$AT(t5, e7, i6) {
    this._$Ct = t5, this._$AM = e7, this._$Ci = i6;
  }
  _$AS(t5, e7) {
    return this.update(t5, e7);
  }
  update(t5, e7) {
    return this.render(...e7);
  }
};

// node_modules/lit-html/directives/style-map.js
var n5 = "important";
var i5 = " !" + n5;
var o5 = e6(class extends i4 {
  constructor(t5) {
    if (super(t5), t5.type !== t4.ATTRIBUTE || "style" !== t5.name || t5.strings?.length > 2)
      throw Error("The `styleMap` directive must be used in the `style` attribute and must be the only part in the attribute.");
  }
  render(t5) {
    return Object.keys(t5).reduce((e7, r7) => {
      const s4 = t5[r7];
      return null == s4 ? e7 : e7 + `${r7 = r7.includes("-") ? r7 : r7.replace(/(?:^(webkit|moz|ms|o)|)(?=[A-Z])/g, "-$&").toLowerCase()}:${s4};`;
    }, "");
  }
  update(e7, [r7]) {
    const { style: s4 } = e7.element;
    if (void 0 === this.ft)
      return this.ft = new Set(Object.keys(r7)), this.render(r7);
    for (const t5 of this.ft)
      null == r7[t5] && (this.ft.delete(t5), t5.includes("-") ? s4.removeProperty(t5) : s4[t5] = null);
    for (const t5 in r7) {
      const e8 = r7[t5];
      if (null != e8) {
        this.ft.add(t5);
        const r8 = "string" == typeof e8 && e8.endsWith(i5);
        t5.includes("-") || r8 ? s4.setProperty(t5, r8 ? e8.slice(0, -11) : e8, r8 ? n5 : "") : s4[t5] = e8;
      }
    }
    return w;
  }
});

// libs/embed-libs/embed/package.json
var version = "5.1.4";

// libs/embed-libs/embed/src/lib/luzmo-service.ts
var LuzmoService = class {
  constructor() {
    this._components = [];
  }
};
var luzmo_service_default = new LuzmoService();

// libs/embed-libs/embed/src/lib/helper/event-bus-handler.ts
var _setEventHandler = () => {
  if (typeof window.__luzmoMessageHandler === "function") {
    return;
  }
  window.__luzmoMessageHandler = (event) => {
    const dashboardComponent = luzmo_service_default._components.find(
      (c4) => event.data.name === c4._iframeId
    );
    if (!dashboardComponent) {
      return;
    }
    switch (event.data?.type) {
      case "load": {
        dashboardComponent.dispatchEvent(
          new CustomEvent("load", { detail: { data: event.data } })
        );
        if (dashboardComponent._loadTimeoutHandle) {
          clearTimeout(dashboardComponent._loadTimeoutHandle);
          dashboardComponent._loadTimeoutHandle = void 0;
        }
        dashboardComponent.iframeLoaded = true;
        dashboardComponent.dimensions = event.data.dimensions;
        dashboardComponent._calculateDimAfterDashboardLoaded();
        break;
      }
      case "error": {
        dashboardComponent.dashboardError = true;
        if (dashboardComponent._loadTimeoutHandle) {
          clearTimeout(dashboardComponent._loadTimeoutHandle);
          dashboardComponent._loadTimeoutHandle = void 0;
        }
        dashboardComponent.iframeLoaded = true;
        dashboardComponent._displayError();
        console.error("Dashboard error: ", event.data.errorMsg);
        break;
      }
      case "data": {
        resolvePromise(
          dashboardComponent,
          event.data.requestId,
          event.data.data
        );
        break;
      }
      case "filters": {
        resolvePromise(
          dashboardComponent,
          event.data.requestId,
          event.data.filters
        );
        break;
      }
      case "getAccessibleDashboards": {
        resolvePromise(
          dashboardComponent,
          event.data.requestId,
          event.data.accessibleDashboards
        );
        break;
      }
      case "dashboardAction": {
        dashboardComponent.dispatchEvent(
          new CustomEvent("dashboardAction", { detail: { data: event.data } })
        );
        break;
      }
      case "export": {
        if (event.data.error) {
          rejectPromise(
            dashboardComponent,
            event.data.requestId,
            event.data
          );
        } else {
          dashboardComponent.dispatchEvent(
            new CustomEvent("exported", { detail: { data: event.data } })
          );
          resolvePromise(
            dashboardComponent,
            event.data.requestId,
            event.data
          );
        }
        break;
      }
      case "changedFilters": {
        dashboardComponent.dispatchEvent(
          new CustomEvent("changedFilters", { detail: { data: event.data } })
        );
        break;
      }
      case "customEvent": {
        dashboardComponent.dispatchEvent(
          new CustomEvent("customEvent", { detail: { data: event.data } })
        );
        break;
      }
      case "chartsRendered":
      case "itemsRendered": {
        event.data.type = "itemsRendered";
        dashboardComponent.dispatchEvent(
          new CustomEvent("itemsRendered", { detail: { data: event.data } })
        );
        break;
      }
      case "setEditMode": {
        if (event.data.error) {
          rejectPromise(
            dashboardComponent,
            event.data.requestId,
            event.data
          );
        } else {
          resolvePromise(
            dashboardComponent,
            event.data.requestId,
            event.data
          );
          dashboardComponent.params.editMode = event.data.editMode;
          dashboardComponent._containerStyle = containerStyles(dashboardComponent);
          if (dashboardComponent.dimensions) {
            _setCalculatedDimensions(dashboardComponent);
          }
        }
        break;
      }
      case "setFilters": {
        event.data.error ? rejectPromise(
          dashboardComponent,
          event.data.requestId,
          event.data
        ) : resolvePromise(
          dashboardComponent,
          event.data.requestId,
          event.data
        );
        break;
      }
      case "setPreview": {
        event.data.error ? rejectPromise(
          dashboardComponent,
          event.data.requestId,
          event.data
        ) : resolvePromise(
          dashboardComponent,
          event.data.requestId,
          event.data
        );
        break;
      }
    }
  };
  window.addEventListener("message", window.__luzmoMessageHandler, false);
};

// libs/embed-libs/embed/src/lib/dashboard-component/dashboard-styles.ts
var dashboardStyles = i`
  :host {
    display: block;
    height: 100%;
  }

  @-webkit-keyframes spin {
    0% {
      -webkit-transform: rotate(0deg);
      transform: rotate(0deg);
    }
    100% {
      -webkit-transform: rotate(360deg);
      transform: rotate(360deg);
    }
  }

  @keyframes spin {
    0% {
      -webkit-transform: rotate(0deg);
      transform: rotate(0deg);
    }
    100% {
      -webkit-transform: rotate(360deg);
      transform: rotate(360deg);
    }
  }
  :host {
    position: relative;
    width: 100%;
  }

  .luzmo-embed-dashboard {
    background: transparent;
    padding: 0;
    border: 0;
  }

  .luzmo-embed-dashboard-ede {
  }

  .luzmo-embed-container-ede {
    width: 100%;
    height: 100%;
  }

  .luzmo-container-loader {
    width: 100%;
    position: relative;
  }

  .luzmo-lc {
    margin: 0 auto;
  }

  .luzmo-loading-message {
    width: 100%;
    height: 20px;
    text-align: center;
    margin-bottom: 16px;
    font-family: Helvetica, Arial, sans-serif;
    font-size: 14px;
  }

  .luzmo-loading-circle {
    box-sizing: border-box;
    margin: 16px auto;
    position: relative;
    text-indent: -9999em;
    animation: spin 0.6s infinite linear;
  }

  .luzmo-loading-circle,
  .luzmo-loading-circle:after {
    border-radius: 50%;
    width: 32px;
    height: 32px;
  }

  .itemEmbedding:after {
    width: 24px;
    height: 24px;
  }

  .luzmo-error {
    background: transparent;
    padding: 48px 0;
    margin: 0 auto;
    width: 80%;
    border: 0;
    text-align: center;
    min-height: 192px;
  }

  .dashboard-error {
    font-size: 30px;
    background: transparent;
    padding: 48px 0;
    margin: 0 auto;
    width: 80%;
    border: 0;
    text-align: center;
    min-height: 192px;
  }

  .luzmo-loader {
    top: 0;
    left: 0;
    width: 100%;
    min-height: 400px;
    position: absolute;
    transition-duration: 0.3s;
  }
`;

// libs/embed-libs/embed/src/lib/dashboard-component/dashboard-base.ts
var LuzmoDashboard = class extends s3 {
  constructor(inputParameters) {
    super();
    this.language = "auto";
    this.screenMode = "auto";
    this.switchScreenModeOnResize = true;
    this.editMode = "view";
    this.experimental = false;
    this.iframeStyle = {};
    this._containerStyle = {};
    this._loaderStyle = {};
    this._lcStyle = {};
    this._loadingCircleStyle = {};
    this._containerLoaderStyle = {};
    this._loaderMsg = "";
    this._itemEmbedding = false;
    this.browserError = false;
    this.browserErrorMsg = "";
    this.dashboardError = false;
    this.browserNotSupported = [];
    this.VERSION = version + "-wc-luzmo";
    this.isLibraryLoaded = false;
    this.currentAuthKey = null;
    this.currentAuthToken = null;
    this.attributesMap = {
      dashboardId: "dashboard-id",
      dashboardSlug: "dashboard-slug",
      itemId: "item-id",
      language: "language",
      qeVersion: "qe-version",
      screenMode: "screen-mode",
      switchScreenModeOnResize: "switch-screen-mode-on-resize",
      authKey: "auth-key",
      authToken: "auth-token",
      editMode: "edit-mode",
      mainColor: "main-color",
      accentColor: "accent-color",
      loaderBackground: "loader-background",
      loaderFontColor: "loader-font-color",
      loaderSpinnerColor: "loader-spinner-color",
      loaderSpinnerBackground: "loader-spinner-background",
      appServer: "app-server",
      apiHost: "api-host",
      itemDimensions: "item-dimensions",
      timezoneId: "timezone-id"
    };
    this.loaderSize = 32;
    this.borderWidth = 3;
    this.minHeight = 400;
    this._containerId = uuidv4();
    this._iframeId = `luzmo-dashboard-${this._containerId}`;
    this.iframeLoaded = false;
    this.promises = {};
    this.setInputParameters(inputParameters);
    this.params = this._setOptions();
    this._onResize = () => {
      if (this.shadowRoot?.querySelector(".dashboard-container")) {
        this.containerWidth = this.luzmoEmbedContainer.offsetWidth;
        if (this.params.switchScreenModeOnResize && this.params.screenMode === "auto") {
          for (const d3 of EXTENTS) {
            if (d3.width[0] <= window.innerWidth && window.innerWidth <= d3.width[1]) {
              this.currentScreenMode = d3.mode;
            }
          }
        }
        if (this.iframeLoaded) {
          this._calculateDimAfterDashboardLoaded();
        } else {
          if (this.dimensions) {
            this.containerWidth = this.luzmoEmbedContainer.offsetWidth;
            _setCalculatedDimensions(this);
          }
        }
      }
    };
    this._resizeObserver = new ResizeObserver((entries) => {
      window.requestAnimationFrame(() => {
        if (!Array.isArray(entries) || entries.length === 0) {
          return;
        }
        this._onResize();
      });
    });
  }
  loadLibrary(appServer) {
    if (remotesMap[appServer]) {
      return Promise.resolve(remotesMap[appServer].module);
    }
    remotesMap[appServer] = {
      url: appServer?.startsWith("http://localhost") ? `${removeTrailingSlash(appServer)}/remoteEntry.mjs` : `${removeTrailingSlash(appServer)}/dashboard-app/remoteEntry.mjs`,
      format: "esm",
      from: "vite",
      module: void 0
    };
    return __federationMethodGetRemote(appServer, "./DashboardApp").then((module2) => {
      const wrappedModule = __federationMethodWrapDefault(module2, true);
      remotesMap[appServer].module = wrappedModule;
      return wrappedModule;
    });
  }
  setInputParameters(inputParameters) {
    if (!inputParameters) {
      return;
    }
    const params = [
      "dashboardId",
      "dashboardSlug",
      "itemId",
      "language",
      "qeVersion",
      "screenMode",
      "switchScreenModeOnResize",
      "accentColor",
      "mainColor",
      "authToken",
      "authKey",
      "editMode",
      "loaderBackground",
      "loaderFontColor",
      "loaderSpinnerColor",
      "loaderSpinnerBackground",
      "timezoneId",
      "itemDimensions",
      "apiHost"
    ];
    for (const key of params) {
      const value = inputParameters[key];
      if (value) {
        this[key] = value;
      }
    }
    if (inputParameters.experimental) {
      this.experimental = inputParameters.experimental;
    }
    if (inputParameters.appServer) {
      this.appServer = addTrailingSlash(inputParameters.appServer);
    }
  }
  // Set resize event listener
  connectedCallback() {
    super.connectedCallback();
  }
  // Set resize event listener
  disconnectedCallback() {
    super.disconnectedCallback();
    this._resizeObserver.disconnect();
    for (const promiseKey of Object.keys(this.promises)) {
      this.promises[promiseKey].reject(
        new Error("Dashboard component was removed")
      );
    }
  }
  firstUpdated() {
    this._loaderMsg = getLoader(this)?.loadingMsg;
    this._loaderStyles();
    luzmo_service_default._components.push(this);
    this.initComponent();
    if (this.luzmoEmbedContainer) {
      this._resizeObserver.observe(this.luzmoEmbedContainer);
    }
  }
  initComponent() {
    this.params = this._setOptions();
    if (this.getIsViewMode()) {
      this.loadLibrary(this.params.appServer).then(async () => {
        this.isLibraryLoaded = true;
        await this.updateComplete;
        const oldReference = this.renderRoot.querySelector("dashboard-component");
        this.dashboardReference = oldReference.cloneNode(true);
        oldReference.replaceWith(this.dashboardReference);
        this.dashboardReference.addEventListener("load", (payload) => {
          this.containerWidth = this.dashboardReference.offsetWidth;
          this.dimensions = payload.detail.dimensions;
          _setCalculatedDimensions(this);
          window.dispatchEvent(new Event("resize"));
          this.dispatchEvent(new CustomEvent("load", payload));
        });
        this.dashboardReference.addEventListener("itemsRendered", (payload) => this.dispatchEvent(new CustomEvent("itemsRendered", payload)));
        this.dashboardReference.addEventListener("customEvent", (payload) => this.dispatchEvent(new CustomEvent("customEvent", payload)));
        this.dashboardReference.addEventListener("changedFilters", (payload) => this.dispatchEvent(new CustomEvent("changedFilters", payload)));
        this.dashboardReference.addEventListener("dashboardAction", (payload) => this.dispatchEvent(new CustomEvent("dashboardAction", payload)));
        this.dashboardReference.addEventListener("export", (payload) => this.dispatchEvent(new CustomEvent("exported", payload)));
        this.setAllAttributes();
      });
    } else {
      this.containerWidth = this.container?.clientWidth;
      _extendAndValidate(this);
      _browserCheck(this);
      this._setupLoader();
      if (!this.params.error) {
        this._loaderStyles();
        _setEventHandler();
        this._iframeUrl = createIframeUrl(this);
        this.iframe?.setAttribute("src", this._iframeUrl);
        this.iframeStyle = calculateIframeStyles(this);
      }
    }
  }
  setAllAttributes() {
    if (this.dashboardReference) {
      for (const key in this.attributesMap) {
        if (!["authKey", "authToken", "apiHost", "itemDimensions"].includes(key) && this.params[key]) {
          this.dashboardReference.setAttribute(this.attributesMap[key], this.params[key]);
        }
        if (this.params.itemDimensions) {
          this.dashboardReference.setAttribute("item-dimensions", JSON.stringify(this.params.itemDimensions));
        }
        if (this.params.apiHost) {
          this.dashboardReference.setAttribute("api-host", removeTrailingSlash(this.params.apiHost));
        }
        if (this.params.key) {
          this.dashboardReference.setAttribute("auth-key", this.params.key);
        }
        if (this.params.token) {
          this.dashboardReference.setAttribute("auth-token", this.params.token);
        }
      }
    }
  }
  // Reload dashboard when triggerproperty changes
  updated(changedProperties) {
    const triggerProperties = /* @__PURE__ */ new Set([
      "dashboardId",
      "dashboardSlug",
      "itemId",
      "language",
      "qeVersion",
      "editMode",
      "screenMode",
      "switchScreenModeOnResize",
      "authToken",
      "authKey",
      "appServer",
      "timezoneId",
      "itemDimensions"
    ]);
    if (this.getIsViewMode()) {
      if (this.dashboardReference) {
        for (const [propertyName] of changedProperties.entries()) {
          if (propertyName === "appSever") {
            this.loadLibrary(changedProperties[propertyName]).then(async () => {
              this.isLibraryLoaded = true;
              await this.updateComplete;
              this.dashboardReference = this.renderRoot.querySelector("dashboard-component");
              this.setAllAttributes();
            });
          } else if (this.attributesMap[propertyName]) {
            const mappedAttributeName = this.attributesMap[propertyName] ?? propertyName;
            const attributeValue = typeof this[propertyName] === "object" ? JSON.stringify(this[propertyName]) : this[propertyName]?.toString() || this[propertyName];
            this.dashboardReference.setAttribute(mappedAttributeName, attributeValue);
          }
        }
      }
    } else {
      for (const [propertyName] of changedProperties.entries()) {
        if (triggerProperties.has(propertyName) && propertyName !== "editMode") {
          this.iframeLoaded = false;
          if (this._loadTimeoutHandle) {
            clearTimeout(this._loadTimeoutHandle);
          }
          this._loadTimeoutHandle = setTimeout(() => {
            console.warn(
              "Failed to load dashboard, appServer is likely down or unreachable."
            );
          }, 15 * 1e3);
          this.params = this._setOptions();
          _extendAndValidate(this);
          this._setupLoader();
          this._loaderStyles();
          this._iframeUrl = createIframeUrl(this);
          if (this.iframe) {
            this._updateIframeUrl();
          }
          this.iframeStyle = calculateIframeStyles(this);
        } else if (triggerProperties.has(propertyName) && propertyName === "editMode") {
          if (this.iframeLoaded) {
            this.setEditMode(this.editMode);
          }
          this.params = this._setOptions();
        }
      }
    }
  }
  // Set params
  _setOptions() {
    const defaultParameters = {
      dashboardId: this.dashboardId,
      dashboardSlug: this.dashboardSlug
    };
    if (this.itemId) {
      defaultParameters.itemId = this.itemId;
      this._itemEmbedding = true;
    } else {
      this._itemEmbedding = false;
    }
    if (this.language) {
      defaultParameters.language = this.language;
    }
    if (this.qeVersion) {
      defaultParameters.qeVersion = this.qeVersion;
    }
    if (this.screenMode) {
      defaultParameters.screenMode = this.screenMode;
    }
    defaultParameters.switchScreenModeOnResize = this.switchScreenModeOnResize;
    if (this.authToken) {
      defaultParameters.token = this.authToken;
    }
    if (this.authKey) {
      defaultParameters.key = this.authKey;
    }
    if (this.editMode) {
      defaultParameters.editMode = this.editMode;
    }
    if (this.mainColor) {
      defaultParameters.mainColor = this.mainColor;
    }
    if (this.accentColor) {
      defaultParameters.accentColor = this.accentColor;
    }
    if (this.loaderBackground) {
      defaultParameters.loaderBackground = this.loaderBackground;
    }
    if (this.loaderFontColor) {
      defaultParameters.loaderFontColor = this.loaderFontColor;
    }
    if (this.loaderSpinnerColor) {
      defaultParameters.loaderSpinnerColor = this.loaderSpinnerColor;
    }
    if (this.loaderSpinnerBackground) {
      defaultParameters.loaderSpinnerBackground = this.loaderSpinnerBackground;
    }
    if (this.timezoneId) {
      defaultParameters.timezoneId = this.timezoneId;
    }
    if (this.itemDimensions) {
      defaultParameters.itemDimensions = this.itemDimensions;
    }
    if (this.appServer) {
      this.appServer = this.getIsViewMode() ? removeTrailingSlash(this.appServer) : addTrailingSlash(this.appServer);
      defaultParameters.appServer = this.appServer;
    } else {
      this.appServer = addTrailingSlash(
        DEFAULT_PARAMETERS.appServer
      );
      defaultParameters.appServer = this.appServer;
    }
    if (this.apiHost) {
      this.apiHost = this.getIsViewMode() ? removeTrailingSlash(this.apiHost) : addTrailingSlash(this.apiHost);
      defaultParameters.apiHost = this.apiHost;
    } else {
      this.apiHost = this.appServer.includes("luzmo.com") ? this.appServer.replace("app", "api") : this.appServer;
      defaultParameters.apiHost = this.apiHost;
    }
    this.currentAuthKey = this.authKey;
    this.currentAuthToken = this.authToken;
    return defaultParameters;
  }
  _updateIframeUrl() {
    updateIframeUrl(this.iframe, this._iframeUrl);
  }
  // Setup loader / browser error
  _setupLoader() {
    const data = getLoader(this);
    this._loaderMsg = data.loadingMsg;
    if (data.errorMsg) {
      this.browserError = true;
      this.browserErrorMsg = data.errorMsg;
      this._displayError();
    }
  }
  _displayError() {
    if (this.dashboardError) {
      this._loaderStyles();
      this.iframeStyle.opacity = 1;
    }
    if (this.browserError) {
      this._containerStyle.display = "none";
    }
  }
  _calculateDimAfterDashboardLoaded() {
    this.dashboardError = false;
    this._loaderStyles();
    _setCalculatedDimensions(this);
  }
  // Set loader styles
  _loaderStyles() {
    this._containerStyle = containerStyles(this);
    this._loaderStyle = loaderStyles(this);
    this._loadingCircleStyle = plStyles(this);
    this._containerLoaderStyle = containerLoaderStyles(this);
    this._lcStyle = lcStyles(this);
  }
  static get styles() {
    return [dashboardStyles];
  }
  render() {
    let toBeRenderedHTML;
    const loader = x`
      <div
        class="luzmo-container-loader"
        id="luzmo-container-loader-${this._containerId}"
        style=${o5(
      this._containerLoaderStyle
    )}
      >
        <div
          class="luzmo-loader"
          id="luzmo-loader-${this._containerId}"
          style=${o5(this._loaderStyle)}
        >
          <div
            class="luzmo-lc"
            style=${o5(this._lcStyle)}
          >
            <div
              class="luzmo-loading-circle"
              style=${o5(
      this._loadingCircleStyle
    )}
            ></div>
            ${this._itemEmbedding ? "" : x`
                <div
                  class="luzmo-loading-message"
                  style=${o5({ color: this.loaderFontColor })}
                >
                  ${this._loaderMsg}
                </div>
              `}
          </div>
        </div>
      </div>
    `;
    if (this.getIsViewMode()) {
      toBeRenderedHTML = this.isLibraryLoaded ? x`
          ${this.dashboardId || this.dashboardSlug ? x`
            <dashboard-component dashboard-id="${this.params.dashboardId}" app-server="${this.params.appServer}"
                                 api-host="${this.params.apiHost}"></dashboard-component>` : x`
            <div id="luzmo-error-${this._containerId}">
              <h3 class="luzmo-error">${this.browserErrorMsg}</h3>
            </div>
          `}
        ` : x`${loader}`;
    } else {
      toBeRenderedHTML = x`
        ${this.dashboardId || this.dashboardSlug ? x`
            ${loader}
            <iframe
              class="luzmo-embed-dashboard ${!this.getIsViewMode() ? `luzmo-embed-dashboard-ede` : ""}"
              id="${this._iframeId}"
              name="${this._iframeId}"
              style=${o5(this.iframeStyle)}
              src="${this._iframeUrl}"
              frameborder="0"
            ></iframe>
          ` : x`<!-- the dashboard will inititialize when you provide a dashboardSlug or a dashboardId -->`}
        ${this.browserError ? x`
            <div id="luzmo-error-${this._containerId}">
              <h3 class="luzmo-error">${this.browserErrorMsg}</h3>
            </div>
          ` : ""}
      `;
    }
    return x`
      <div class="${luzmoEmbedContainerClass} ${!this.getIsViewMode() ? `${luzmoEmbedContainerClass}-ede` : ""}">
        <div
          class="dashboard-container"
          style=${o5(Object.assign({}, this._containerStyle, {
      width: this.iframeStyle.width ?? "100%",
      height: this.iframeStyle.height ?? "100%"
    }))}
        >
          ${toBeRenderedHTML}
        </div>
      </div>

    `;
  }
  getDashboards() {
    return new Promise(
      (resolve) => resolve(luzmo_service_default._components)
    );
  }
  // Public methods
  getData(itemId) {
    if (this.getIsViewMode() && this.dashboardReference) {
      return this.dashboardReference.getData(itemId);
    }
    if (!this.getIsViewMode()) {
      return getData(this, this.iframe, itemId);
    }
    return Promise.reject();
  }
  getFilters() {
    if (this.getIsViewMode() && this.dashboardReference) {
      return this.dashboardReference.getFilters();
    }
    if (!this.getIsViewMode()) {
      return getFilters(this, this.iframe);
    }
    return Promise.reject();
  }
  setFilters(itemId, filters) {
    if (this.getIsViewMode() && this.dashboardReference) {
      return this.dashboardReference.setFilters(itemId, filters);
    }
    if (!this.getIsViewMode()) {
      return setFilters(this, this.iframe, itemId, filters);
    }
    return Promise.reject();
  }
  setAuthorization(key, token) {
    this.currentAuthKey = key;
    this.currentAuthToken = token;
    if (this.getIsViewMode() && this.dashboardReference) {
      return this.dashboardReference.setAuthorization(key, token);
    }
    if (!this.getIsViewMode()) {
      return setAuthorization(this, this.iframe, key, token);
    }
    return Promise.reject();
  }
  refreshData(itemId) {
    if (this.getIsViewMode() && this.dashboardReference) {
      return this.dashboardReference.refreshData(itemId);
    }
    if (!this.getIsViewMode()) {
      return itemId ? refreshData(this, this.iframe, itemId) : refreshData(this, this.iframe);
    }
    return Promise.reject();
  }
  reloadDashboard() {
    if (this.getIsViewMode() && this.dashboardReference) {
      return this.dashboardReference.reloadDashboard();
    }
    if (!this.getIsViewMode()) {
      return reloadDashboard(this, this.iframe);
    }
    return Promise.reject();
  }
  exportDashboard(format) {
    if (this.getIsViewMode() && this.dashboardReference) {
      return this.dashboardReference.exportDashboard(format);
    }
    if (!this.getIsViewMode()) {
      return format ? exportDashboard(this, this.iframe, format) : exportDashboard(this, this.iframe);
    }
    return Promise.reject();
  }
  getAccessibleDashboards() {
    return getAccessibleDashboards(this);
  }
  addFilters(itemId, filters) {
    return this.dashboardReference.addFilters(itemId, filters);
  }
  setEditMode(editMode) {
    if (this.dashboardReference && editMode === "view") {
      this.editMode = "view";
      this.params.editMode = "view";
      this.initComponent();
      return Promise.resolve({
        type: "setEditMode",
        editMode
      });
    }
    if (this.dashboardReference && ["editFull", "editLimited"].includes(editMode)) {
      this.editMode = editMode;
      this.params.editMode = editMode;
      this.initComponent();
      return Promise.resolve({
        type: "setEditMode",
        editMode
      });
    }
    if (this.getIsViewMode() && this.dashboardReference) {
      return this.dashboardReference.setEditMode(editMode);
    }
    if (!this.getIsViewMode()) {
      return setEditMode(this, this.iframe, editMode);
    }
    return Promise.reject();
  }
  setPreview(preview) {
    if (this.getIsViewMode() && this.dashboardReference) {
      return this.dashboardReference.setPreview(preview);
    }
    return setPreview(this, this.iframe, preview);
  }
  getIsViewMode() {
    return this.experimental && this.params.editMode === "view" || this.params.editMode === void 0;
  }
};
__decorateClass([
  e4(".dashboard-container")
], LuzmoDashboard.prototype, "container", 2);
__decorateClass([
  e4(`.${luzmoEmbedContainerClass}`)
], LuzmoDashboard.prototype, "luzmoEmbedContainer", 2);
__decorateClass([
  e4("iframe")
], LuzmoDashboard.prototype, "iframe", 2);
__decorateClass([
  n3()
], LuzmoDashboard.prototype, "dashboardId", 2);
__decorateClass([
  n3()
], LuzmoDashboard.prototype, "dashboardSlug", 2);
__decorateClass([
  n3()
], LuzmoDashboard.prototype, "itemId", 2);
__decorateClass([
  n3()
], LuzmoDashboard.prototype, "language", 2);
__decorateClass([
  n3()
], LuzmoDashboard.prototype, "qeVersion", 2);
__decorateClass([
  n3()
], LuzmoDashboard.prototype, "screenMode", 2);
__decorateClass([
  n3({
    converter: {
      fromAttribute: (value) => value === "true",
      toAttribute: String
    }
  })
], LuzmoDashboard.prototype, "switchScreenModeOnResize", 2);
__decorateClass([
  n3()
], LuzmoDashboard.prototype, "editMode", 2);
__decorateClass([
  n3()
], LuzmoDashboard.prototype, "mainColor", 2);
__decorateClass([
  n3()
], LuzmoDashboard.prototype, "accentColor", 2);
__decorateClass([
  n3()
], LuzmoDashboard.prototype, "authToken", 2);
__decorateClass([
  n3()
], LuzmoDashboard.prototype, "authKey", 2);
__decorateClass([
  n3()
], LuzmoDashboard.prototype, "loaderBackground", 2);
__decorateClass([
  n3()
], LuzmoDashboard.prototype, "loaderFontColor", 2);
__decorateClass([
  n3()
], LuzmoDashboard.prototype, "loaderSpinnerColor", 2);
__decorateClass([
  n3()
], LuzmoDashboard.prototype, "loaderSpinnerBackground", 2);
__decorateClass([
  n3()
], LuzmoDashboard.prototype, "appServer", 2);
__decorateClass([
  n3()
], LuzmoDashboard.prototype, "timezoneId", 2);
__decorateClass([
  n3()
], LuzmoDashboard.prototype, "apiHost", 2);
__decorateClass([
  n3({ type: Object })
], LuzmoDashboard.prototype, "itemDimensions", 2);
__decorateClass([
  n3({
    converter: {
      fromAttribute: (value) => value === "true",
      toAttribute: String
    }
  })
], LuzmoDashboard.prototype, "experimental", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "_containerId", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "_iframeId", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "params", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "_iframeUrl", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "iframeLoaded", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "promises", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "iframeStyle", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "_containerStyle", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "_loaderStyle", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "_lcStyle", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "_loadingCircleStyle", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "_containerLoaderStyle", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "_loaderMsg", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "containerWidth", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "dimensions", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "currentScreenMode", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "_itemEmbedding", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "browserError", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "browserErrorMsg", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "dashboardError", 2);
__decorateClass([
  n3({ attribute: false })
], LuzmoDashboard.prototype, "browserNotSupported", 2);
__decorateClass([
  n3()
], LuzmoDashboard.prototype, "VERSION", 2);
__decorateClass([
  r4()
], LuzmoDashboard.prototype, "isLibraryLoaded", 2);
__decorateClass([
  r4()
], LuzmoDashboard.prototype, "dashboardReference", 2);
__decorateClass([
  r4()
], LuzmoDashboard.prototype, "currentAuthKey", 2);
__decorateClass([
  r4()
], LuzmoDashboard.prototype, "currentAuthToken", 2);

// libs/embed-libs/embed/src/lib/dashboard-component/dashboard-web-component.ts
var LuzmoDashboardWebComponent = class extends LuzmoDashboard {
};
LuzmoDashboardWebComponent = __decorateClass([
  t("luzmo-dashboard-web-component")
], LuzmoDashboardWebComponent);
/*! Bundled license information:

@lit/reactive-element/decorators/custom-element.js:
  (**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   *)

@lit/reactive-element/css-tag.js:
  (**
   * @license
   * Copyright 2019 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   *)

@lit/reactive-element/reactive-element.js:
  (**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   *)

@lit/reactive-element/decorators/property.js:
  (**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   *)

@lit/reactive-element/decorators/state.js:
  (**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   *)

@lit/reactive-element/decorators/event-options.js:
  (**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   *)

@lit/reactive-element/decorators/base.js:
  (**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   *)

@lit/reactive-element/decorators/query.js:
  (**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   *)

@lit/reactive-element/decorators/query-all.js:
  (**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   *)

@lit/reactive-element/decorators/query-async.js:
  (**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   *)

@lit/reactive-element/decorators/query-assigned-elements.js:
  (**
   * @license
   * Copyright 2021 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   *)

@lit/reactive-element/decorators/query-assigned-nodes.js:
  (**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   *)

lit-html/lit-html.js:
  (**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   *)

lit-element/lit-element.js:
  (**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   *)

lit-html/is-server.js:
  (**
   * @license
   * Copyright 2022 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   *)

lit-html/directive.js:
  (**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   *)

lit-html/directives/style-map.js:
  (**
   * @license
   * Copyright 2018 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   *)
*/
if (typeof module.exports == "object" && typeof exports == "object") {
  var __cp = (to, from, except, desc) => {
    if ((from && typeof from === "object") || typeof from === "function") {
      for (let key of Object.getOwnPropertyNames(from)) {
        if (!Object.prototype.hasOwnProperty.call(to, key) && key !== except)
        Object.defineProperty(to, key, {
          get: () => from[key],
          enumerable: !(desc = Object.getOwnPropertyDescriptor(from, key)) || desc.enumerable,
        });
      }
    }
    return to;
  };
  module.exports = __cp(module.exports, exports);
}
return module.exports;
}))
